<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>DMS_designer package &mdash; DMS Designer v0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="DMS_designer" href="modules.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            DMS Designer
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Example Notebooks:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="notebooks.html">Example Notebooks</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">DMS_designer</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">DMS_designer package</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-DMS_designer.library_analyzer">DMS_designer.library_analyzer module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-DMS_designer.library_maker">DMS_designer.library_maker module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#classes">Classes</a></li>
<li class="toctree-l4"><a class="reference internal" href="#functions">Functions</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-DMS_designer">Module contents</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">DMS Designer</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="modules.html">DMS_designer</a></li>
      <li class="breadcrumb-item active">DMS_designer package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/DMS_designer.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="dms-designer-package">
<h1>DMS_designer package<a class="headerlink" href="#dms-designer-package" title="Permalink to this heading"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this heading"></a></h2>
</section>
<section id="module-DMS_designer.library_analyzer">
<span id="dms-designer-library-analyzer-module"></span><h2>DMS_designer.library_analyzer module<a class="headerlink" href="#module-DMS_designer.library_analyzer" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.count_over_thresh">
<span class="sig-prename descclassname"><span class="pre">DMS_designer.library_analyzer.</span></span><span class="sig-name descname"><span class="pre">count_over_thresh</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#count_over_thresh"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.count_over_thresh" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.make_density">
<span class="sig-prename descclassname"><span class="pre">DMS_designer.library_analyzer.</span></span><span class="sig-name descname"><span class="pre">make_density</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#make_density"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.make_density" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.line_plotter">
<span class="sig-prename descclassname"><span class="pre">DMS_designer.library_analyzer.</span></span><span class="sig-name descname"><span class="pre">line_plotter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#line_plotter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.line_plotter" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.correspondance_plotter">
<span class="sig-prename descclassname"><span class="pre">DMS_designer.library_analyzer.</span></span><span class="sig-name descname"><span class="pre">correspondance_plotter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#correspondance_plotter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.correspondance_plotter" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.density_scatter">
<span class="sig-prename descclassname"><span class="pre">DMS_designer.library_analyzer.</span></span><span class="sig-name descname"><span class="pre">density_scatter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#density_scatter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.density_scatter" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.OLS_plotter">
<span class="sig-prename descclassname"><span class="pre">DMS_designer.library_analyzer.</span></span><span class="sig-name descname"><span class="pre">OLS_plotter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ols_model</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#OLS_plotter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.OLS_plotter" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.gini">
<span class="sig-prename descclassname"><span class="pre">DMS_designer.library_analyzer.</span></span><span class="sig-name descname"><span class="pre">gini</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sorted_arr</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#gini"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.gini" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.multi_gini">
<span class="sig-prename descclassname"><span class="pre">DMS_designer.library_analyzer.</span></span><span class="sig-name descname"><span class="pre">multi_gini</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">xs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Library</span> <span class="pre">skew'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scatter_kwargs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#multi_gini"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.multi_gini" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.depth_plot">
<span class="sig-prename descclassname"><span class="pre">DMS_designer.library_analyzer.</span></span><span class="sig-name descname"><span class="pre">depth_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">xs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Library</span> <span class="pre">depth'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#depth_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.depth_plot" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.make_ROC">
<span class="sig-prename descclassname"><span class="pre">DMS_designer.library_analyzer.</span></span><span class="sig-name descname"><span class="pre">make_ROC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">true_class</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metrics</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">make_plots</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_hist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sorting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#make_ROC"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.make_ROC" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.thresh_collapse">
<span class="sig-prename descclassname"><span class="pre">DMS_designer.library_analyzer.</span></span><span class="sig-name descname"><span class="pre">thresh_collapse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-</span> <span class="pre">0.5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#thresh_collapse"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.thresh_collapse" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">DMS_designer.library_analyzer.</span></span><span class="sig-name descname"><span class="pre">Library</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_cols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_cols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rate_method</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alphabet</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">WT_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fitness_measure</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_cols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_cols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rate_method</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alphabet</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">WT_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fitness_measure</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.__init__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.save"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.save" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.load">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.load"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.load" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.build_from_reads">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">build_from_reads</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_cols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_cols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rate_method</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alphabet</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">WT_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">process_call</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name_regex</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sheet_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.build_from_reads"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.build_from_reads" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.merge_by_overlap">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">merge_by_overlap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">library_dest</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">library_source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dest_position_shift</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_position_shift</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_WT</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Position'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">automatically_rescale_linearly</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">automatically_rescale_by_controls</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'AA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_val</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['WT',</span> <span class="pre">'*']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.merge_by_overlap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.merge_by_overlap" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.fitness_frame">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">fitness_frame</span></span><a class="headerlink" href="#DMS_designer.library_analyzer.Library.fitness_frame" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.rep_average_frame">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">rep_average_frame</span></span><a class="headerlink" href="#DMS_designer.library_analyzer.Library.rep_average_frame" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.group_average_frame">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">group_average_frame</span></span><a class="headerlink" href="#DMS_designer.library_analyzer.Library.group_average_frame" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.total_average_frame">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">total_average_frame</span></span><a class="headerlink" href="#DMS_designer.library_analyzer.Library.total_average_frame" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.check_skew">
<span class="sig-name descname"><span class="pre">check_skew</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sns_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'notebook'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rc_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Library</span> <span class="pre">skew'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.check_skew"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.check_skew" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.check_depth">
<span class="sig-name descname"><span class="pre">check_depth</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sns_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'notebook'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rc_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Read</span> <span class="pre">depth'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.check_depth"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.check_depth" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.check_barcode_distribution_skew">
<span class="sig-name descname"><span class="pre">check_barcode_distribution_skew</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sns_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'notebook'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rc_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Read</span> <span class="pre">depth'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.check_barcode_distribution_skew"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.check_barcode_distribution_skew" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.check_barcode_distribution_depth">
<span class="sig-name descname"><span class="pre">check_barcode_distribution_depth</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sns_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'notebook'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rc_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Read</span> <span class="pre">depth'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.check_barcode_distribution_depth"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.check_barcode_distribution_depth" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.check_controls">
<span class="sig-name descname"><span class="pre">check_controls</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">control_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">positive_vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">negative_vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sns_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'notebook'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rc_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plt_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'strip'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_hist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.check_controls"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.check_controls" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.make_replicate_checks">
<span class="sig-name descname"><span class="pre">make_replicate_checks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sns_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'notebook'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rc_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">grid_kwargs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.make_replicate_checks"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.make_replicate_checks" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.get_rescaler_by_replicate">
<span class="sig-name descname"><span class="pre">get_rescaler_by_replicate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dest_replicate</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_replicate</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.get_rescaler_by_replicate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.get_rescaler_by_replicate" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.rescale_by_replicate">
<span class="sig-name descname"><span class="pre">rescale_by_replicate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dest_replicate</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_replicate</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.rescale_by_replicate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.rescale_by_replicate" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.get_rescaler_by_replicate_groups">
<span class="sig-name descname"><span class="pre">get_rescaler_by_replicate_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dest_replicate</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_replicate</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.get_rescaler_by_replicate_groups"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.get_rescaler_by_replicate_groups" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.rescale_by_replicate_group">
<span class="sig-name descname"><span class="pre">rescale_by_replicate_group</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dest_replicate</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_replicate</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.rescale_by_replicate_group"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.rescale_by_replicate_group" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.get_multi_replicate_rescaler_by_groups">
<span class="sig-name descname"><span class="pre">get_multi_replicate_rescaler_by_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dest_replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_replicates</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.get_multi_replicate_rescaler_by_groups"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.get_multi_replicate_rescaler_by_groups" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.rescale_multiple_replicates_by_group">
<span class="sig-name descname"><span class="pre">rescale_multiple_replicates_by_group</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dest_replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_replicates</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.rescale_multiple_replicates_by_group"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.rescale_multiple_replicates_by_group" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.get_multi_replicate_rescaler_by_controls">
<span class="sig-name descname"><span class="pre">get_multi_replicate_rescaler_by_controls</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dest_replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.get_multi_replicate_rescaler_by_controls"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.get_multi_replicate_rescaler_by_controls" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.rescale_multiple_replicates_by_controls">
<span class="sig-name descname"><span class="pre">rescale_multiple_replicates_by_controls</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dest_replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.rescale_multiple_replicates_by_controls"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.rescale_multiple_replicates_by_controls" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.make_initial_skew_checks">
<span class="sig-name descname"><span class="pre">make_initial_skew_checks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sns_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'notebook'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rc_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.make_initial_skew_checks"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.make_initial_skew_checks" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.get_single_mutant_mutability_score">
<span class="sig-name descname"><span class="pre">get_single_mutant_mutability_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">control_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">WT_vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">thresh</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">invert</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.get_single_mutant_mutability_score"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.get_single_mutant_mutability_score" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.get_single_mutant_property_enrichments">
<span class="sig-name descname"><span class="pre">get_single_mutant_property_enrichments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">property_sets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">thresh</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Position'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">char_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'AA'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.get_single_mutant_property_enrichments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.get_single_mutant_property_enrichments" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.draw_singles_heatmap">
<span class="sig-name descname"><span class="pre">draw_singles_heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mut_num_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">WT_vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">grid</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tick_interval</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_pos</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.draw_singles_heatmap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.draw_singles_heatmap" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.draw_mutability_score">
<span class="sig-name descname"><span class="pre">draw_mutability_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">WT_vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">thresh</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">invert_mutability</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">grid</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_pos</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.draw_mutability_score"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.draw_mutability_score" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.draw_property_scores">
<span class="sig-name descname"><span class="pre">draw_property_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">property_sets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">thresh</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">grid</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_pos</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.draw_property_scores"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.draw_property_scores" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.draw_feature_sets">
<span class="sig-name descname"><span class="pre">draw_feature_sets</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">feature_sets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">grid</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_pos</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_multi_rows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.draw_feature_sets"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.draw_feature_sets" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_analyzer.Library.make_singles_heatmap">
<span class="sig-name descname"><span class="pre">make_singles_heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">mut_num_col='Mut_num',</span> <span class="pre">position_col='Position',</span> <span class="pre">main_cmap='RdBu',</span> <span class="pre">mutability_cmap=&lt;matplotlib.colors.LinearSegmentedColormap</span> <span class="pre">object&gt;,</span> <span class="pre">bad_color='xkcd:grey',</span> <span class="pre">sns_context='notebook',</span> <span class="pre">rc_params={},</span> <span class="pre">grid=False,</span> <span class="pre">invert_mutability=False,</span> <span class="pre">norm=None,</span> <span class="pre">seq_start=0,</span> <span class="pre">tick_interval=5,</span> <span class="pre">plot_pos=None,</span> <span class="pre">plot_mutability=False,</span> <span class="pre">plot_properties=False,</span> <span class="pre">plot_features=False,</span> <span class="pre">mutability_norm=None,</span> <span class="pre">property_norm=None,</span> <span class="pre">property_sets={'Charged':</span> <span class="pre">['R',</span> <span class="pre">'H',</span> <span class="pre">'K',</span> <span class="pre">'D',</span> <span class="pre">'E'],</span> <span class="pre">'Hydrophobic':</span> <span class="pre">['A',</span> <span class="pre">'V',</span> <span class="pre">'I',</span> <span class="pre">'L',</span> <span class="pre">'M',</span> <span class="pre">'F',</span> <span class="pre">'Y',</span> <span class="pre">'W'],</span> <span class="pre">'Polar':</span> <span class="pre">['S',</span> <span class="pre">'T',</span> <span class="pre">'N',</span> <span class="pre">'Q']},</span> <span class="pre">property_cmap='RdBu',</span> <span class="pre">feature_sets=[],</span> <span class="pre">control_col='AA',</span> <span class="pre">WT_vals=['WT'],</span> <span class="pre">stop_vals=['*'],</span> <span class="pre">thresh=0,</span> <span class="pre">allow_multi_rows=False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.make_singles_heatmap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_analyzer.Library.make_singles_heatmap" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

</section>
<section id="module-DMS_designer.library_maker">
<span id="dms-designer-library-maker-module"></span><h2>DMS_designer.library_maker module<a class="headerlink" href="#module-DMS_designer.library_maker" title="Permalink to this heading"></a></h2>
<p>Library Maker for Protein Mutagenesis Studies</p>
<p>This module provides tools for creating oligonucleotide libraries for deep mutational
scanning and protein mutagenesis studies. It includes functionality for generating
single and multiple amino acid mutants, creating synonymous nucleotide variants,
and analyzing library composition.</p>
<p>The module reads codon usage data from ‘codon_usage.xlsx’ to generate realistic
nucleotide sequences that encode the desired protein variants.</p>
<section id="classes">
<h3>Classes<a class="headerlink" href="#classes" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>LibraryMaker</dt><dd><p>Main class for creating mutagenesis libraries</p>
</dd>
</dl>
</section>
<section id="functions">
<h3>Functions<a class="headerlink" href="#functions" title="Permalink to this heading"></a></h3>
<p>aa_to_codon : Convert amino acid to codon using codon usage table
protein_to_orf : Convert protein sequence to open reading frame
orf_to_protein : Convert open reading frame to protein sequence
summarize_file : Generate analysis plots and statistics for library files</p>
<div class="admonition-examples admonition">
<p class="admonition-title">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># Create a library with single and double mutants</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lm</span> <span class="o">=</span> <span class="n">LibraryMaker</span><span class="p">(</span><span class="n">wt_nt</span><span class="o">=</span><span class="s2">&quot;ATGGCCGAA&quot;</span><span class="p">,</span> <span class="n">single_mutants</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">doubles_to_make</span><span class="o">=</span><span class="s1">&#39;equal&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Analyze a library file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">summarize_file</span><span class="p">(</span><span class="s1">&#39;my_library.csv&#39;</span><span class="p">,</span> <span class="n">out_path</span><span class="o">=</span><span class="s1">&#39;./analysis/&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<dl class="py function">
<dt class="sig sig-object py" id="DMS_designer.library_maker.aa_to_codon">
<span class="sig-prename descclassname"><span class="pre">DMS_designer.library_maker.</span></span><span class="sig-name descname"><span class="pre">aa_to_codon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">aa</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_to_codon_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'*':</span> <span class="pre">array(['TAA',</span> <span class="pre">'TGA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'A':</span> <span class="pre">array(['GCG',</span> <span class="pre">'GCC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'C':</span> <span class="pre">array(['TGC',</span> <span class="pre">'TGT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'D':</span> <span class="pre">array(['GAT',</span> <span class="pre">'GAC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'E':</span> <span class="pre">array(['GAA',</span> <span class="pre">'GAG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'F':</span> <span class="pre">array(['TTT',</span> <span class="pre">'TTC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'G':</span> <span class="pre">array(['GGC',</span> <span class="pre">'GGT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'H':</span> <span class="pre">array(['CAT',</span> <span class="pre">'CAC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'I':</span> <span class="pre">array(['ATT',</span> <span class="pre">'ATC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'K':</span> <span class="pre">array(['AAA',</span> <span class="pre">'AAG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'L':</span> <span class="pre">array(['CTG',</span> <span class="pre">'TTA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'M':</span> <span class="pre">array(['ATG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'N':</span> <span class="pre">array(['AAC',</span> <span class="pre">'AAT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'P':</span> <span class="pre">array(['CCG',</span> <span class="pre">'CCA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'Q':</span> <span class="pre">array(['CAG',</span> <span class="pre">'CAA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'R':</span> <span class="pre">array(['CGT',</span> <span class="pre">'CGC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'S':</span> <span class="pre">array(['AGC',</span> <span class="pre">'TCT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'T':</span> <span class="pre">array(['ACC',</span> <span class="pre">'ACG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'V':</span> <span class="pre">array(['GTG',</span> <span class="pre">'GTT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'W':</span> <span class="pre">array(['TGG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'Y':</span> <span class="pre">array(['TAT',</span> <span class="pre">'TAC'],</span> <span class="pre">dtype=object)}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rng</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#aa_to_codon"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_maker.aa_to_codon" title="Permalink to this definition"></a></dt>
<dd><p>Convert an amino acid to a codon according to intrinsic probabilities.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>aa</strong> (<em>str</em>) – Single amino acid character to convert to codon</p></li>
<li><p><strong>aa_to_codon_dict</strong> (<em>dict</em><em>, </em><em>optional</em>) – Dictionary mapping amino acids to their codons. If None, uses the default codon usage table.</p></li>
<li><p><strong>rng</strong> (<em>numpy.random.Generator</em><em>, </em><em>optional</em>) – Random number generator instance. If None, creates a new default generator.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Three-nucleotide codon encoding the amino acid</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>AssertionError</strong> – If the amino acid is not found in the codon usage table</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Uses the codon usage probabilities defined in the loaded ‘codon_usage.xlsx’ file.
Only codons marked as ‘use=True’ in the table are considered.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="DMS_designer.library_maker.protein_to_orf">
<span class="sig-prename descclassname"><span class="pre">DMS_designer.library_maker.</span></span><span class="sig-name descname"><span class="pre">protein_to_orf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">protein</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rng</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_to_codon_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'*':</span> <span class="pre">array(['TAA',</span> <span class="pre">'TGA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'A':</span> <span class="pre">array(['GCG',</span> <span class="pre">'GCC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'C':</span> <span class="pre">array(['TGC',</span> <span class="pre">'TGT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'D':</span> <span class="pre">array(['GAT',</span> <span class="pre">'GAC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'E':</span> <span class="pre">array(['GAA',</span> <span class="pre">'GAG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'F':</span> <span class="pre">array(['TTT',</span> <span class="pre">'TTC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'G':</span> <span class="pre">array(['GGC',</span> <span class="pre">'GGT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'H':</span> <span class="pre">array(['CAT',</span> <span class="pre">'CAC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'I':</span> <span class="pre">array(['ATT',</span> <span class="pre">'ATC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'K':</span> <span class="pre">array(['AAA',</span> <span class="pre">'AAG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'L':</span> <span class="pre">array(['CTG',</span> <span class="pre">'TTA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'M':</span> <span class="pre">array(['ATG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'N':</span> <span class="pre">array(['AAC',</span> <span class="pre">'AAT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'P':</span> <span class="pre">array(['CCG',</span> <span class="pre">'CCA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'Q':</span> <span class="pre">array(['CAG',</span> <span class="pre">'CAA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'R':</span> <span class="pre">array(['CGT',</span> <span class="pre">'CGC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'S':</span> <span class="pre">array(['AGC',</span> <span class="pre">'TCT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'T':</span> <span class="pre">array(['ACC',</span> <span class="pre">'ACG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'V':</span> <span class="pre">array(['GTG',</span> <span class="pre">'GTT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'W':</span> <span class="pre">array(['TGG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'Y':</span> <span class="pre">array(['TAT',</span> <span class="pre">'TAC'],</span> <span class="pre">dtype=object)}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#protein_to_orf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_maker.protein_to_orf" title="Permalink to this definition"></a></dt>
<dd><p>Convert a protein sequence to an open reading frame using codon usage table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein</strong> (<em>str</em>) – Protein sequence as a string of amino acid characters</p></li>
<li><p><strong>rng</strong> (<em>numpy.random.Generator</em><em>, </em><em>optional</em>) – Random number generator instance. If None, creates a new default generator.</p></li>
<li><p><strong>aa_to_codon_dict</strong> (<em>dict</em><em>, </em><em>optional</em>) – Dictionary mapping amino acids to their codons. If None, uses the default codon usage table.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Nucleotide sequence encoding the protein according to codon usage probabilities</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Each amino acid is converted to a codon using the probabilities defined in
the codon usage table. Multiple calls with the same protein may produce
different nucleotide sequences due to synonymous codon variation.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="DMS_designer.library_maker.orf_to_protein">
<span class="sig-prename descclassname"><span class="pre">DMS_designer.library_maker.</span></span><span class="sig-name descname"><span class="pre">orf_to_protein</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">orf,</span> <span class="pre">codon_df=</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">aa</span>&#160; <span class="pre">frac_of_aa</span>&#160; <span class="pre">ct_per_k</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">ct</span>&#160;&#160;&#160; <span class="pre">use</span> <span class="pre">codon</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">TAA</span>&#160;&#160;&#160; <span class="pre">*</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.61</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">2.0</span>&#160;&#160;&#160; <span class="pre">7356</span>&#160;&#160; <span class="pre">True</span> <span class="pre">TGA</span>&#160;&#160;&#160; <span class="pre">*</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.30</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">1.0</span>&#160;&#160;&#160; <span class="pre">3623</span>&#160;&#160; <span class="pre">True</span> <span class="pre">TAG</span>&#160;&#160;&#160; <span class="pre">*</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.09</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.3</span>&#160;&#160;&#160;&#160; <span class="pre">989</span>&#160; <span class="pre">False</span> <span class="pre">GCG</span>&#160;&#160;&#160; <span class="pre">A</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.33</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">30.1</span>&#160; <span class="pre">110308</span>&#160;&#160; <span class="pre">True</span> <span class="pre">GCC</span>&#160;&#160;&#160; <span class="pre">A</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.26</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">24.2</span>&#160;&#160; <span class="pre">88721</span>&#160;&#160; <span class="pre">True</span> <span class="pre">...</span>&#160;&#160; <span class="pre">..</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">...</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">...</span>&#160;&#160;&#160;&#160; <span class="pre">...</span>&#160;&#160;&#160; <span class="pre">...</span> <span class="pre">GTC</span>&#160;&#160;&#160; <span class="pre">V</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.20</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">14.3</span>&#160;&#160; <span class="pre">52439</span>&#160; <span class="pre">False</span> <span class="pre">GTA</span>&#160;&#160;&#160; <span class="pre">V</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.17</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">11.6</span>&#160;&#160; <span class="pre">42420</span>&#160; <span class="pre">False</span> <span class="pre">TGG</span>&#160;&#160;&#160; <span class="pre">W</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">1.00</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">13.9</span>&#160;&#160; <span class="pre">50991</span>&#160;&#160; <span class="pre">True</span> <span class="pre">TAT</span>&#160;&#160;&#160; <span class="pre">Y</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.59</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">17.5</span>&#160;&#160; <span class="pre">63937</span>&#160;&#160; <span class="pre">True</span> <span class="pre">TAC</span>&#160;&#160;&#160; <span class="pre">Y</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.41</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">12.2</span>&#160;&#160; <span class="pre">44631</span>&#160;&#160; <span class="pre">True</span>&#160; <span class="pre">[64</span> <span class="pre">rows</span> <span class="pre">x</span> <span class="pre">5</span> <span class="pre">columns]</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#orf_to_protein"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_maker.orf_to_protein" title="Permalink to this definition"></a></dt>
<dd><p>Convert an open reading frame to a protein sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>orf</strong> (<em>str</em>) – Nucleotide sequence to translate. Must be a multiple of 3.</p></li>
<li><p><strong>codon_df</strong> (<em>pd.DataFrame</em><em>, </em><em>optional</em>) – DataFrame containing codon usage information. If None, uses the default codon usage table.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Protein sequence translated from the nucleotide sequence</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>AssertionError</strong> – If the nucleotide sequence length is not a multiple of 3</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Uses the codon-to-amino acid mapping defined in the loaded ‘codon_usage.xlsx’ file.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="DMS_designer.library_maker.summarize_file">
<span class="sig-prename descclassname"><span class="pre">DMS_designer.library_maker.</span></span><span class="sig-name descname"><span class="pre">summarize_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'./'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fig_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">12</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'talk'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#summarize_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_maker.summarize_file" title="Permalink to this definition"></a></dt>
<dd><p>Generate comprehensive analysis and visualization of a mutagenesis library file.</p>
<p>This function reads a CSV file containing library variants and generates multiple
plots and statistics including variant counts, mutant distributions, heatmaps,
and position-by-position analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> (<em>str</em>) – Path to the CSV file containing the library data. Expected columns:
‘name’, ‘protein’, ‘orf’, ‘oligo’</p></li>
<li><p><strong>out_path</strong> (<em>str</em><em>, </em><em>default './'</em>) – Directory path where output figures will be saved</p></li>
<li><p><strong>fig_width</strong> (<em>int</em><em>, </em><em>default 12</em>) – Base width for generated figures. Height is typically equal to width</p></li>
<li><p><strong>context</strong> (<em>{'talk'</em><em>, </em><em>'paper'</em><em>, </em><em>'notebook'</em><em>, </em><em>'poster'}</em><em>, </em><em>default 'talk'</em>) – Seaborn plotting context for figure styling</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>The function generates the following output files:
- variant_count_[filename].pdf: Count of variants by mutation number
- mutant_count_[filename].pdf: Count of unique mutants by mutation number
- mutant_fraction_[filename].pdf: Fraction of possible mutants generated
- mutant_observations_[filename].pdf: Distribution of observations per mutant
- [N]pt_heatmap_[filename].pdf: Heatmaps showing amino acid frequencies at each position
- [N]pt_mutant_heatmap_[filename].pdf: Heatmaps for unique mutants only
- 2pt_2d_heatmap_[filename].pdf: Position-by-position matrix for double mutants</p>
<p>The function also prints summary statistics to the console including:
- Wild-type sequence information
- Counts of variants by mutation number
- Total library size
- Fraction of possible mutants generated</p>
</div>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="DMS_designer.library_maker.LibraryMaker">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">DMS_designer.library_maker.</span></span><span class="sig-name descname"><span class="pre">LibraryMaker</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">wt_nt,</span> <span class="pre">single_mutants=True,</span> <span class="pre">doubles_to_make='equal',</span> <span class="pre">name='Mutant_lib',</span> <span class="pre">var_num=3,</span> <span class="pre">wt_num=100,</span> <span class="pre">max_tries=100,</span> <span class="pre">stop_num=None,</span> <span class="pre">focus_positions=[],</span> <span class="pre">focus_to_make=0,</span> <span class="pre">focus_mutants_per=3,</span> <span class="pre">focus_mutants_per_spread=0,</span> <span class="pre">focus_mutants_per_min=0,</span> <span class="pre">focus_mutants_per_max=0,</span> <span class="pre">upstream_dna='ATGGAAAGTGTCCCAGGAGACTACAGC',</span> <span class="pre">downstream_dna='CAGACAAGTAACTCAAGACTTAATGGC',</span> <span class="pre">out_path='./',</span> <span class="pre">seed=None,</span> <span class="pre">codon_df=</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">aa</span>&#160; <span class="pre">frac_of_aa</span>&#160; <span class="pre">ct_per_k</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">ct</span>&#160;&#160;&#160; <span class="pre">use</span> <span class="pre">codon</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">TAA</span>&#160;&#160;&#160; <span class="pre">*</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.61</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">2.0</span>&#160;&#160;&#160; <span class="pre">7356</span>&#160;&#160; <span class="pre">True</span> <span class="pre">TGA</span>&#160;&#160;&#160; <span class="pre">*</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.30</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">1.0</span>&#160;&#160;&#160; <span class="pre">3623</span>&#160;&#160; <span class="pre">True</span> <span class="pre">TAG</span>&#160;&#160;&#160; <span class="pre">*</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.09</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.3</span>&#160;&#160;&#160;&#160; <span class="pre">989</span>&#160; <span class="pre">False</span> <span class="pre">GCG</span>&#160;&#160;&#160; <span class="pre">A</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.33</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">30.1</span>&#160; <span class="pre">110308</span>&#160;&#160; <span class="pre">True</span> <span class="pre">GCC</span>&#160;&#160;&#160; <span class="pre">A</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.26</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">24.2</span>&#160;&#160; <span class="pre">88721</span>&#160;&#160; <span class="pre">True</span> <span class="pre">...</span>&#160;&#160; <span class="pre">..</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">...</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">...</span>&#160;&#160;&#160;&#160; <span class="pre">...</span>&#160;&#160;&#160; <span class="pre">...</span> <span class="pre">GTC</span>&#160;&#160;&#160; <span class="pre">V</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.20</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">14.3</span>&#160;&#160; <span class="pre">52439</span>&#160; <span class="pre">False</span> <span class="pre">GTA</span>&#160;&#160;&#160; <span class="pre">V</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.17</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">11.6</span>&#160;&#160; <span class="pre">42420</span>&#160; <span class="pre">False</span> <span class="pre">TGG</span>&#160;&#160;&#160; <span class="pre">W</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">1.00</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">13.9</span>&#160;&#160; <span class="pre">50991</span>&#160;&#160; <span class="pre">True</span> <span class="pre">TAT</span>&#160;&#160;&#160; <span class="pre">Y</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.59</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">17.5</span>&#160;&#160; <span class="pre">63937</span>&#160;&#160; <span class="pre">True</span> <span class="pre">TAC</span>&#160;&#160;&#160; <span class="pre">Y</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.41</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">12.2</span>&#160;&#160; <span class="pre">44631</span>&#160;&#160; <span class="pre">True</span>&#160; <span class="pre">[64</span> <span class="pre">rows</span> <span class="pre">x</span> <span class="pre">5</span> <span class="pre">columns],</span> <span class="pre">aa_to_codon_dict={'*':</span> <span class="pre">array(['TAA',</span> <span class="pre">'TGA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'A':</span> <span class="pre">array(['GCG',</span> <span class="pre">'GCC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'C':</span> <span class="pre">array(['TGC',</span> <span class="pre">'TGT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'D':</span> <span class="pre">array(['GAT',</span> <span class="pre">'GAC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'E':</span> <span class="pre">array(['GAA',</span> <span class="pre">'GAG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'F':</span> <span class="pre">array(['TTT',</span> <span class="pre">'TTC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'G':</span> <span class="pre">array(['GGC',</span> <span class="pre">'GGT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'H':</span> <span class="pre">array(['CAT',</span> <span class="pre">'CAC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'I':</span> <span class="pre">array(['ATT',</span> <span class="pre">'ATC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'K':</span> <span class="pre">array(['AAA',</span> <span class="pre">'AAG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'L':</span> <span class="pre">array(['CTG',</span> <span class="pre">'TTA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'M':</span> <span class="pre">array(['ATG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'N':</span> <span class="pre">array(['AAC',</span> <span class="pre">'AAT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'P':</span> <span class="pre">array(['CCG',</span> <span class="pre">'CCA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'Q':</span> <span class="pre">array(['CAG',</span> <span class="pre">'CAA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'R':</span> <span class="pre">array(['CGT',</span> <span class="pre">'CGC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'S':</span> <span class="pre">array(['AGC',</span> <span class="pre">'TCT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'T':</span> <span class="pre">array(['ACC',</span> <span class="pre">'ACG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'V':</span> <span class="pre">array(['GTG',</span> <span class="pre">'GTT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'W':</span> <span class="pre">array(['TGG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'Y':</span> <span class="pre">array(['TAT',</span> <span class="pre">'TAC'],</span> <span class="pre">dtype=object)},</span> <span class="pre">fig_width=8,</span> <span class="pre">context='talk',</span> <span class="pre">fig_path=None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#LibraryMaker"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_maker.LibraryMaker" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Create oligonucleotide libraries for protein mutagenesis studies.</p>
<p>This class generates comprehensive mutagenesis libraries including single and
multiple amino acid mutants, synonymous nucleotide variants, and focused
mutagenesis at specific positions. The library is designed for deep mutational
scanning experiments.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wt_nt</strong> (<em>str</em>) – The nucleotide sequence of the wild-type variant. This sequence should
encode an amino acid sequence, start on the first nucleotide of the first
codon to be mutagenized, and end on the last nucleotide of the last
codon to be mutagenized.</p></li>
<li><p><strong>single_mutants</strong> (<em>bool</em><em>, </em><em>default True</em>) – If True, generate every possible single amino acid mutant of the wild-type
variant.</p></li>
<li><p><strong>doubles_to_make</strong> (<em>str</em><em>, </em><em>float</em><em>, or </em><em>int</em><em>, </em><em>default 'equal'</em>) – <p>Controls the number of double mutants to generate:</p>
<ul>
<li><p>’equal’: Generate random double mutants equal to the number of single mutants</p></li>
<li><p>’max:N’: Generate as many random double mutants as possible without exceeding
N total oligos for synthesis</p></li>
<li><p>float: Generate that fraction of the number of single mutants as random
double mutants</p></li>
<li><p>int: Generate exactly that number of random double mutants</p></li>
</ul>
</p></li>
<li><p><strong>name</strong> (<em>str</em><em>, </em><em>default 'Mutant_lib'</em>) – Name of the library to be created. Used as prefix for output files.</p></li>
<li><p><strong>var_num</strong> (<em>int</em><em>, </em><em>default 3</em>) – Number of different synonymous nucleotide variants to generate for each mutant.</p></li>
<li><p><strong>wt_num</strong> (<em>int</em><em>, </em><em>default 100</em>) – Number of different synonymous nucleotide variants to generate for the
wild-type sequence.</p></li>
<li><p><strong>max_tries</strong> (<em>int</em><em>, </em><em>default 100</em>) – Maximum number of attempts to generate a unique random mutant or variant
before giving up.</p></li>
<li><p><strong>focus_positions</strong> (<em>list of int</em><em>, </em><em>default</em><em> [</em><em>]</em>) – Amino acid positions (0-indexed) to be mutated in greater depth.</p></li>
<li><p><strong>focus_to_make</strong> (<em>int</em><em>, </em><em>float</em><em>, or </em><em>str</em><em>, </em><em>default 0</em>) – <p>Controls focused mutagenesis when focus_positions is specified:</p>
<ul>
<li><p>int/float: Generate that many random combinations of mutations at focus positions</p></li>
<li><p>’singles’: Generate all single mutants of the focus positions</p></li>
<li><p>’doubles’: Generate all double mutants of the focus positions</p></li>
<li><p>’singles_doubles’: Generate all single and double mutants of the focus positions</p></li>
<li><p>’all’: Generate every possible combination of mutations at the focus positions</p></li>
</ul>
</p></li>
<li><p><strong>focus_mutants_per</strong> (<em>int</em><em>, </em><em>default 3</em>) – Average number of mutations desired per focused mutant (used when focus_to_make
is numeric).</p></li>
<li><p><strong>focus_mutants_per_spread</strong> (<em>float</em><em>, </em><em>default 0</em>) – Probability spread for mutation numbers other than focus_mutants_per.
Probability decays arithmetically as you move away from the target.</p></li>
<li><p><strong>focus_mutants_per_min</strong> (<em>int</em><em>, </em><em>default 0</em>) – Minimum number of mutations per focused mutant.</p></li>
<li><p><strong>focus_mutants_per_max</strong> (<em>int</em><em>, </em><em>default 0</em>) – Maximum number of mutations per focused mutant.</p></li>
<li><p><strong>upstream_dna</strong> (<em>str</em><em>, </em><em>default 'ATGGAAAGTGTCCCAGGAGACTACAGC'</em>) – Upstream adapter sequence to append to the 5’ end of oligonucleotides.</p></li>
<li><p><strong>downstream_dna</strong> (<em>str</em><em>, </em><em>default 'CAGACAAGTAACTCAAGACTTAATGGC'</em>) – Downstream adapter sequence to append to the 3’ end of oligonucleotides.</p></li>
<li><p><strong>out_path</strong> (<em>str</em><em>, </em><em>default './'</em>) – Directory path for output files. Directory must already exist.</p></li>
<li><p><strong>seed</strong> (<em>int</em><em>, </em><em>optional</em>) – Random seed for reproducibility. If None, uses system default.</p></li>
<li><p><strong>fig_width</strong> (<em>int</em><em>, </em><em>default 8</em>) – Base width for generated analysis figures.</p></li>
<li><p><strong>context</strong> (<em>{'talk'</em><em>, </em><em>'paper'</em><em>, </em><em>'notebook'</em><em>, </em><em>'poster'}</em><em>, </em><em>default 'talk'</em>) – Seaborn plotting context for figure styling.</p></li>
</ul>
</dd>
<dt class="field-even">Variables<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>wt_nt</strong> (<em>str</em>) – Wild-type nucleotide sequence</p></li>
<li><p><strong>wt_protein</strong> (<em>str</em>) – Wild-type protein sequence</p></li>
<li><p><strong>L</strong> (<em>int</em>) – Length of the protein sequence</p></li>
<li><p><strong>variant_dict</strong> (<em>dict</em>) – Dictionary mapping variant names to protein sequences</p></li>
<li><p><strong>orf_dict</strong> (<em>dict</em>) – Dictionary mapping variant names to nucleotide sequences</p></li>
<li><p><strong>orf_df</strong> (<em>pandas.DataFrame</em>) – Final library data with columns: name, protein, orf, oligo</p></li>
<li><p><strong>file_name</strong> (<em>str</em>) – Name of the saved CSV file</p></li>
<li><p><strong>date_time</strong> (<em>str</em>) – Timestamp of library creation</p></li>
</ul>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>The library generation process:
1. Creates single mutants (if enabled)
2. Creates random double mutants (if enabled)
3. Creates focused mutants (if focus positions specified)
4. Generates synonymous nucleotide variants for each protein variant
5. Saves results to CSV file
6. Generates analysis plots and statistics</p>
</div>
<div class="admonition-examples admonition">
<p class="admonition-title">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># Create a simple library with single mutants only</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lm</span> <span class="o">=</span> <span class="n">LibraryMaker</span><span class="p">(</span><span class="s2">&quot;ATGGCCGAA&quot;</span><span class="p">,</span> <span class="n">single_mutants</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">doubles_to_make</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Create a comprehensive library with focused mutagenesis</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lm</span> <span class="o">=</span> <span class="n">LibraryMaker</span><span class="p">(</span><span class="s2">&quot;ATGGCCGAA&quot;</span><span class="p">,</span>
<span class="gp">... </span>                  <span class="n">single_mutants</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">... </span>                  <span class="n">doubles_to_make</span><span class="o">=</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span>
<span class="gp">... </span>                  <span class="n">focus_positions</span><span class="o">=</span><span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">15</span><span class="p">],</span>
<span class="gp">... </span>                  <span class="n">focus_to_make</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>This class creates oligo libraries for mutagenesis studies of protein function</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wt_nt</strong> (<em>string</em>) – The nucleotide sequence of the WT variant. This sequence should encode an AA sequence,
start on the first NT of the first codon to be mutagenized, and end on the last nt of the last codon to be mutagenized</p></li>
<li><p><strong>single_mutants</strong> (<em>bool Default is True</em>) – If true, make every possible single AA mutant of the WT variant</p></li>
<li><p><strong>doubles_to_make</strong> (<em>string</em><em>, </em><em>float</em><em>, or </em><em>int Default is 'equal</em>) – If this is the string ‘equal’ make random double mutants equal to the number of single mutants made
If this is a string starting with ‘max:’ follwed by an int, make as many random double mutants as possible
without allowing the number of oligos required to synthesize single and double mutants to exceed the number listed
If this is a float, Make that fraction of the number of single mutants as random duble mutants
If this is an int, Make that number of random double mutants</p></li>
<li><p><strong>name</strong> (<em>string Default is 'Mutant_lib'</em>) – name of the library to be created, will be appended as a prefix to the output files</p></li>
<li><p><strong>var_num</strong> (<em>int default 3</em>) – The number of different synonymous nt variants to be made for each mutant</p></li>
<li><p><strong>wt_num</strong> (<em>int default 100</em>) – The number of different synsonymous nt vatriants to be made for the WT sequence</p></li>
<li><p><strong>max_tries</strong> (<em>int default 100</em>) – The number of times to retry making a random mutant or variant if an already existing sequence was created</p></li>
<li><p><strong>stop_num</strong> (<em>int default None</em>) – The number of different synonymous nt variants to be made for each nonsense variant,
if None, default to var_num</p></li>
<li><p><strong>focus_positions</strong> (<em>list default</em><em> [</em><em>]</em>) – Amino acid positions of AAs to be mutated in greater depth</p></li>
<li><p><strong>focus_to_make</strong> (<em>int</em><em>, </em><em>float</em><em>, or </em><em>{'singles'</em><em>, </em><em>'doubles'</em><em>, </em><em>'singles_doubles'</em><em>, </em><em>'all'} default 0</em>) – This parameter is only used of focus_positions contains at least 1 element
If this is an int or a float, make that many random combinations of mutations at the focus positions
If this is the string ‘singles’, make all single mutants of the focus positions
If this is the string ‘doubles’, make all double mutants of the focus positions
If this is the string ‘singles_doubles’, make all single and double muatants of the focus positions
If this is the string ‘all’, make every combination of mutations at the focus positions</p></li>
<li><p><strong>focus_mutants_per</strong> (<em>int default 0</em>) – This parameter is only used of focus_positions contains at least 1 element and focus_to_make is an int or float &gt;0
this indicates the average number of mutations desired per mutant</p></li>
<li><p><strong>focus_mutants_per_spread</strong> (<em>int default 0</em>) – This parameter is only used of focus_positions contains at least 1 element and focus_to_make is an int or float &gt;0
This is the probability of producing a mutation number other than focus_mutants_per, mutation probability will be evenly distributed
between higher and lower mutation numbers with arithmatic decay as you move away from focus_mutants_per</p></li>
<li><p><strong>focus_mutants_per_min</strong> – This parameter is only used of focus_positions contains at least 1 element and focus_to_make is an int or float &gt;0
This is the minimum mutation number you want to see per mutant</p></li>
<li><p><strong>focus_mutants_per_max</strong> – This parameter is only used of focus_positions contains at least 1 element and focus_to_make is an int or float &gt;0
This is the Maximum mutation number you want to see per mutant</p></li>
<li><p><strong>upstream_dna</strong> (<em>string default 'ATGGAAAGTGTCCCAGGAGACTACAGC'</em>) – This is the upstream adaptor sequence to append to the 5’ end of oligos</p></li>
<li><p><strong>downstream_dna</strong> (<em>string default 'CAGACAAGTAACTCAAGACTTAATGGC'</em>) – This is the downstream adaptor sequence to append to the 3’ end of oligos</p></li>
<li><p><strong>out_path</strong> (<em>string default './'</em>) – path to place output files. This directory must already exist</p></li>
<li><p><strong>codon_df</strong> (<em>pd.DataFrame</em><em>, </em><em>optional</em>) – DataFrame containing codon usage information. If None, uses the default codon usage table.</p></li>
<li><p><strong>aa_to_codon_dict</strong> (<em>dict</em><em>, </em><em>optional</em>) – Dictionary mapping amino acids to their codons. If None, uses the default codon usage table.</p></li>
<li><p><strong>seed</strong> (<em>int Default None</em>) – Random seed. Set to ensure reproducibility.</p></li>
<li><p><strong>fig_width</strong> (<em>int default 8</em>) – Size parameter for output figures</p></li>
<li><p><strong>context</strong> (<em>{'talk'</em><em>, </em><em>'paper'</em><em>, </em><em>'notebook'</em><em>, </em><em>'poster'} default 'talk'</em>) – Used to set plotting configuration. passed to seaborn.set_context</p></li>
<li><p><strong>fig_path</strong> (<em>str</em><em>, </em><em>default None</em>) – Path to save the output figures. If default to out_path.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_maker.LibraryMaker.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">wt_nt,</span> <span class="pre">single_mutants=True,</span> <span class="pre">doubles_to_make='equal',</span> <span class="pre">name='Mutant_lib',</span> <span class="pre">var_num=3,</span> <span class="pre">wt_num=100,</span> <span class="pre">max_tries=100,</span> <span class="pre">stop_num=None,</span> <span class="pre">focus_positions=[],</span> <span class="pre">focus_to_make=0,</span> <span class="pre">focus_mutants_per=3,</span> <span class="pre">focus_mutants_per_spread=0,</span> <span class="pre">focus_mutants_per_min=0,</span> <span class="pre">focus_mutants_per_max=0,</span> <span class="pre">upstream_dna='ATGGAAAGTGTCCCAGGAGACTACAGC',</span> <span class="pre">downstream_dna='CAGACAAGTAACTCAAGACTTAATGGC',</span> <span class="pre">out_path='./',</span> <span class="pre">seed=None,</span> <span class="pre">codon_df=</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">aa</span>&#160; <span class="pre">frac_of_aa</span>&#160; <span class="pre">ct_per_k</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">ct</span>&#160;&#160;&#160; <span class="pre">use</span> <span class="pre">codon</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">TAA</span>&#160;&#160;&#160; <span class="pre">*</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.61</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">2.0</span>&#160;&#160;&#160; <span class="pre">7356</span>&#160;&#160; <span class="pre">True</span> <span class="pre">TGA</span>&#160;&#160;&#160; <span class="pre">*</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.30</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">1.0</span>&#160;&#160;&#160; <span class="pre">3623</span>&#160;&#160; <span class="pre">True</span> <span class="pre">TAG</span>&#160;&#160;&#160; <span class="pre">*</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.09</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.3</span>&#160;&#160;&#160;&#160; <span class="pre">989</span>&#160; <span class="pre">False</span> <span class="pre">GCG</span>&#160;&#160;&#160; <span class="pre">A</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.33</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">30.1</span>&#160; <span class="pre">110308</span>&#160;&#160; <span class="pre">True</span> <span class="pre">GCC</span>&#160;&#160;&#160; <span class="pre">A</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.26</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">24.2</span>&#160;&#160; <span class="pre">88721</span>&#160;&#160; <span class="pre">True</span> <span class="pre">...</span>&#160;&#160; <span class="pre">..</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">...</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">...</span>&#160;&#160;&#160;&#160; <span class="pre">...</span>&#160;&#160;&#160; <span class="pre">...</span> <span class="pre">GTC</span>&#160;&#160;&#160; <span class="pre">V</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.20</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">14.3</span>&#160;&#160; <span class="pre">52439</span>&#160; <span class="pre">False</span> <span class="pre">GTA</span>&#160;&#160;&#160; <span class="pre">V</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.17</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">11.6</span>&#160;&#160; <span class="pre">42420</span>&#160; <span class="pre">False</span> <span class="pre">TGG</span>&#160;&#160;&#160; <span class="pre">W</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">1.00</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">13.9</span>&#160;&#160; <span class="pre">50991</span>&#160;&#160; <span class="pre">True</span> <span class="pre">TAT</span>&#160;&#160;&#160; <span class="pre">Y</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.59</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">17.5</span>&#160;&#160; <span class="pre">63937</span>&#160;&#160; <span class="pre">True</span> <span class="pre">TAC</span>&#160;&#160;&#160; <span class="pre">Y</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.41</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">12.2</span>&#160;&#160; <span class="pre">44631</span>&#160;&#160; <span class="pre">True</span>&#160; <span class="pre">[64</span> <span class="pre">rows</span> <span class="pre">x</span> <span class="pre">5</span> <span class="pre">columns],</span> <span class="pre">aa_to_codon_dict={'*':</span> <span class="pre">array(['TAA',</span> <span class="pre">'TGA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'A':</span> <span class="pre">array(['GCG',</span> <span class="pre">'GCC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'C':</span> <span class="pre">array(['TGC',</span> <span class="pre">'TGT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'D':</span> <span class="pre">array(['GAT',</span> <span class="pre">'GAC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'E':</span> <span class="pre">array(['GAA',</span> <span class="pre">'GAG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'F':</span> <span class="pre">array(['TTT',</span> <span class="pre">'TTC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'G':</span> <span class="pre">array(['GGC',</span> <span class="pre">'GGT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'H':</span> <span class="pre">array(['CAT',</span> <span class="pre">'CAC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'I':</span> <span class="pre">array(['ATT',</span> <span class="pre">'ATC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'K':</span> <span class="pre">array(['AAA',</span> <span class="pre">'AAG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'L':</span> <span class="pre">array(['CTG',</span> <span class="pre">'TTA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'M':</span> <span class="pre">array(['ATG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'N':</span> <span class="pre">array(['AAC',</span> <span class="pre">'AAT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'P':</span> <span class="pre">array(['CCG',</span> <span class="pre">'CCA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'Q':</span> <span class="pre">array(['CAG',</span> <span class="pre">'CAA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'R':</span> <span class="pre">array(['CGT',</span> <span class="pre">'CGC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'S':</span> <span class="pre">array(['AGC',</span> <span class="pre">'TCT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'T':</span> <span class="pre">array(['ACC',</span> <span class="pre">'ACG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'V':</span> <span class="pre">array(['GTG',</span> <span class="pre">'GTT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'W':</span> <span class="pre">array(['TGG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'Y':</span> <span class="pre">array(['TAT',</span> <span class="pre">'TAC'],</span> <span class="pre">dtype=object)},</span> <span class="pre">fig_width=8,</span> <span class="pre">context='talk',</span> <span class="pre">fig_path=None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#LibraryMaker.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_maker.LibraryMaker.__init__" title="Permalink to this definition"></a></dt>
<dd><p>This class creates oligo libraries for mutagenesis studies of protein function</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wt_nt</strong> (<em>string</em>) – The nucleotide sequence of the WT variant. This sequence should encode an AA sequence,
start on the first NT of the first codon to be mutagenized, and end on the last nt of the last codon to be mutagenized</p></li>
<li><p><strong>single_mutants</strong> (<em>bool Default is True</em>) – If true, make every possible single AA mutant of the WT variant</p></li>
<li><p><strong>doubles_to_make</strong> (<em>string</em><em>, </em><em>float</em><em>, or </em><em>int Default is 'equal</em>) – If this is the string ‘equal’ make random double mutants equal to the number of single mutants made
If this is a string starting with ‘max:’ follwed by an int, make as many random double mutants as possible
without allowing the number of oligos required to synthesize single and double mutants to exceed the number listed
If this is a float, Make that fraction of the number of single mutants as random duble mutants
If this is an int, Make that number of random double mutants</p></li>
<li><p><strong>name</strong> (<em>string Default is 'Mutant_lib'</em>) – name of the library to be created, will be appended as a prefix to the output files</p></li>
<li><p><strong>var_num</strong> (<em>int default 3</em>) – The number of different synonymous nt variants to be made for each mutant</p></li>
<li><p><strong>wt_num</strong> (<em>int default 100</em>) – The number of different synsonymous nt vatriants to be made for the WT sequence</p></li>
<li><p><strong>max_tries</strong> (<em>int default 100</em>) – The number of times to retry making a random mutant or variant if an already existing sequence was created</p></li>
<li><p><strong>stop_num</strong> (<em>int default None</em>) – The number of different synonymous nt variants to be made for each nonsense variant,
if None, default to var_num</p></li>
<li><p><strong>focus_positions</strong> (<em>list default</em><em> [</em><em>]</em>) – Amino acid positions of AAs to be mutated in greater depth</p></li>
<li><p><strong>focus_to_make</strong> (<em>int</em><em>, </em><em>float</em><em>, or </em><em>{'singles'</em><em>, </em><em>'doubles'</em><em>, </em><em>'singles_doubles'</em><em>, </em><em>'all'} default 0</em>) – This parameter is only used of focus_positions contains at least 1 element
If this is an int or a float, make that many random combinations of mutations at the focus positions
If this is the string ‘singles’, make all single mutants of the focus positions
If this is the string ‘doubles’, make all double mutants of the focus positions
If this is the string ‘singles_doubles’, make all single and double muatants of the focus positions
If this is the string ‘all’, make every combination of mutations at the focus positions</p></li>
<li><p><strong>focus_mutants_per</strong> (<em>int default 0</em>) – This parameter is only used of focus_positions contains at least 1 element and focus_to_make is an int or float &gt;0
this indicates the average number of mutations desired per mutant</p></li>
<li><p><strong>focus_mutants_per_spread</strong> (<em>int default 0</em>) – This parameter is only used of focus_positions contains at least 1 element and focus_to_make is an int or float &gt;0
This is the probability of producing a mutation number other than focus_mutants_per, mutation probability will be evenly distributed
between higher and lower mutation numbers with arithmatic decay as you move away from focus_mutants_per</p></li>
<li><p><strong>focus_mutants_per_min</strong> – This parameter is only used of focus_positions contains at least 1 element and focus_to_make is an int or float &gt;0
This is the minimum mutation number you want to see per mutant</p></li>
<li><p><strong>focus_mutants_per_max</strong> – This parameter is only used of focus_positions contains at least 1 element and focus_to_make is an int or float &gt;0
This is the Maximum mutation number you want to see per mutant</p></li>
<li><p><strong>upstream_dna</strong> (<em>string default 'ATGGAAAGTGTCCCAGGAGACTACAGC'</em>) – This is the upstream adaptor sequence to append to the 5’ end of oligos</p></li>
<li><p><strong>downstream_dna</strong> (<em>string default 'CAGACAAGTAACTCAAGACTTAATGGC'</em>) – This is the downstream adaptor sequence to append to the 3’ end of oligos</p></li>
<li><p><strong>out_path</strong> (<em>string default './'</em>) – path to place output files. This directory must already exist</p></li>
<li><p><strong>codon_df</strong> (<em>pd.DataFrame</em><em>, </em><em>optional</em>) – DataFrame containing codon usage information. If None, uses the default codon usage table.</p></li>
<li><p><strong>aa_to_codon_dict</strong> (<em>dict</em><em>, </em><em>optional</em>) – Dictionary mapping amino acids to their codons. If None, uses the default codon usage table.</p></li>
<li><p><strong>seed</strong> (<em>int Default None</em>) – Random seed. Set to ensure reproducibility.</p></li>
<li><p><strong>fig_width</strong> (<em>int default 8</em>) – Size parameter for output figures</p></li>
<li><p><strong>context</strong> (<em>{'talk'</em><em>, </em><em>'paper'</em><em>, </em><em>'notebook'</em><em>, </em><em>'poster'} default 'talk'</em>) – Used to set plotting configuration. passed to seaborn.set_context</p></li>
<li><p><strong>fig_path</strong> (<em>str</em><em>, </em><em>default None</em>) – Path to save the output figures. If default to out_path.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_maker.LibraryMaker.create_focused_singles">
<span class="sig-name descname"><span class="pre">create_focused_singles</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#LibraryMaker.create_focused_singles"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_maker.LibraryMaker.create_focused_singles" title="Permalink to this definition"></a></dt>
<dd><p>Create all single amino acid mutants at focused positions.</p>
<p>Generates every possible single amino acid substitution at each position
specified in focus_positions, excluding the wild-type amino acid.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary mapping variant names to protein sequences for focused
single mutants. Names follow pattern 1pt_PPAA where PP is the
position (01-indexed) and AA is the amino acid.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Updates self.variant_dict with the new variants.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_maker.LibraryMaker.create_focused_doubles">
<span class="sig-name descname"><span class="pre">create_focused_doubles</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#LibraryMaker.create_focused_doubles"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_maker.LibraryMaker.create_focused_doubles" title="Permalink to this definition"></a></dt>
<dd><p>Create all double amino acid mutants at focused positions.</p>
<p>Generates every possible combination of two amino acid substitutions
at pairs of positions specified in focus_positions, excluding wild-type
amino acids and stop codons.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary mapping variant names to protein sequences for focused
double mutants. Names follow pattern 2pt_PP1AA1_PP2AA2 where PP
is the position (01-indexed) and AA is the amino acid.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Updates self.variant_dict with the new variants.
Excludes stop codons (*) from possible substitutions.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_maker.LibraryMaker.create_focused_all_combinations">
<span class="sig-name descname"><span class="pre">create_focused_all_combinations</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#LibraryMaker.create_focused_all_combinations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_maker.LibraryMaker.create_focused_all_combinations" title="Permalink to this definition"></a></dt>
<dd><p>Create all possible combinations of mutations at focused positions.</p>
<p>Generates every possible combination of amino acid substitutions at the
focused positions, from single mutations up to mutations at all focused
positions simultaneously. Excludes wild-type amino acids and stop codons.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary mapping variant names to protein sequences for all focused
mutant combinations. Names follow pattern <a href="#id1"><span class="problematic" id="id2">Npt_PP1AA1_PP2AA2_</span></a>…
where N is the number of mutations, PP is position (01-indexed),
and AA is the amino acid.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Updates self.variant_dict with the new variants.
Excludes stop codons (*) from possible substitutions.
Generates combinations of all sizes from 1 to len(focus_positions).</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_maker.LibraryMaker.create_focused_random_combinations">
<span class="sig-name descname"><span class="pre">create_focused_random_combinations</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#LibraryMaker.create_focused_random_combinations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_maker.LibraryMaker.create_focused_random_combinations" title="Permalink to this definition"></a></dt>
<dd><p>Create random combinations of mutations at focused positions.</p>
<p>Generates random combinations of amino acid substitutions at focused
positions according to specified probability distributions and constraints.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary mapping variant names to protein sequences for random
focused mutant combinations. Names follow pattern <a href="#id3"><span class="problematic" id="id4">Npt_PP1AA1_PP2AA2_</span></a>…
where N is the number of mutations, PP is position (01-indexed),
and AA is the amino acid.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Updates self.variant_dict with the new variants.</p>
<p>The number of mutations per variant is controlled by:
- focus_mutants_per: Target average number of mutations
- focus_mutants_per_spread: Probability spread around the target
- focus_mutants_per_min: Minimum mutations per variant
- focus_mutants_per_max: Maximum mutations per variant</p>
<p>If spread &gt; 0, probability decays arithmetically as you move away
from the target number of mutations.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_maker.LibraryMaker.create_singles">
<span class="sig-name descname"><span class="pre">create_singles</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#LibraryMaker.create_singles"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_maker.LibraryMaker.create_singles" title="Permalink to this definition"></a></dt>
<dd><p>Create all possible single amino acid mutants.</p>
<p>Generates every possible single amino acid substitution at each position
in the protein sequence, excluding the wild-type amino acid.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary mapping variant names to protein sequences for single
mutants. Names follow pattern 1pt_PPAA where PP is the position
(01-indexed) and AA is the amino acid.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Updates self.variant_dict with the new variants.
Creates (L * 20) single mutants where L is the protein length.
The 20 includes the stop codons.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_maker.LibraryMaker.create_doubles">
<span class="sig-name descname"><span class="pre">create_doubles</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#LibraryMaker.create_doubles"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_maker.LibraryMaker.create_doubles" title="Permalink to this definition"></a></dt>
<dd><p>Create random double amino acid mutants.</p>
<p>Generates random combinations of two amino acid substitutions at different
positions in the protein sequence. Uses a tracking matrix to avoid
generating duplicate combinations.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary mapping variant names to protein sequences for random
double mutants. Names follow pattern 2pt_PP1AA1_PP2AA2 where PP
is the position (01-indexed) and AA is the amino acid.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Updates self.variant_dict with the new variants.
Uses self.already_seen matrix to track which combinations have been
generated to avoid duplicates.
Excludes stop codons (*) from possible substitutions.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_maker.LibraryMaker.generate_nt_seqs">
<span class="sig-name descname"><span class="pre">generate_nt_seqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">aa_to_codon_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'*':</span> <span class="pre">array(['TAA',</span> <span class="pre">'TGA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'A':</span> <span class="pre">array(['GCG',</span> <span class="pre">'GCC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'C':</span> <span class="pre">array(['TGC',</span> <span class="pre">'TGT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'D':</span> <span class="pre">array(['GAT',</span> <span class="pre">'GAC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'E':</span> <span class="pre">array(['GAA',</span> <span class="pre">'GAG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'F':</span> <span class="pre">array(['TTT',</span> <span class="pre">'TTC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'G':</span> <span class="pre">array(['GGC',</span> <span class="pre">'GGT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'H':</span> <span class="pre">array(['CAT',</span> <span class="pre">'CAC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'I':</span> <span class="pre">array(['ATT',</span> <span class="pre">'ATC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'K':</span> <span class="pre">array(['AAA',</span> <span class="pre">'AAG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'L':</span> <span class="pre">array(['CTG',</span> <span class="pre">'TTA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'M':</span> <span class="pre">array(['ATG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'N':</span> <span class="pre">array(['AAC',</span> <span class="pre">'AAT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'P':</span> <span class="pre">array(['CCG',</span> <span class="pre">'CCA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'Q':</span> <span class="pre">array(['CAG',</span> <span class="pre">'CAA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'R':</span> <span class="pre">array(['CGT',</span> <span class="pre">'CGC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'S':</span> <span class="pre">array(['AGC',</span> <span class="pre">'TCT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'T':</span> <span class="pre">array(['ACC',</span> <span class="pre">'ACG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'V':</span> <span class="pre">array(['GTG',</span> <span class="pre">'GTT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'W':</span> <span class="pre">array(['TGG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'Y':</span> <span class="pre">array(['TAT',</span> <span class="pre">'TAC'],</span> <span class="pre">dtype=object)}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#LibraryMaker.generate_nt_seqs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_maker.LibraryMaker.generate_nt_seqs" title="Permalink to this definition"></a></dt>
<dd><p>Generate synonymous nucleotide sequences for all protein variants.</p>
<p>Creates multiple synonymous nucleotide sequences for each protein variant
using the codon usage table. The number of variants depends on whether
it’s the wild-type (wt_num variants) or a mutant (var_num variants).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>aa_to_codon_dict</strong> (<em>dict</em><em>, </em><em>optional</em>) – Dictionary mapping amino acids to their codons. If None, uses the default codon usage table.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary mapping variant names to nucleotide sequences.
Names follow pattern PROTEIN_NAME_vN where N is the variant number.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Updates self.orf_dict with the nucleotide sequences.</p>
<p>For wild-type variants, generates wt_num synonymous sequences.
For mutant variants, generates var_num synonymous sequences.</p>
<p>The first wild-type variant uses the exact input nucleotide sequence.
Subsequent variants are generated by randomly selecting one of the two most common codons for each aa.</p>
<p>Attempts up to max_tries times to generate unique sequences before
giving up on a particular variant.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.library_maker.LibraryMaker.save_file">
<span class="sig-name descname"><span class="pre">save_file</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#LibraryMaker.save_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_maker.LibraryMaker.save_file" title="Permalink to this definition"></a></dt>
<dd><p>Save the library to a CSV file and generate oligonucleotide sequences.</p>
<p>Creates a pandas DataFrame with all library variants and saves it to
a CSV file. The DataFrame includes protein sequences, nucleotide
sequences, and full oligonucleotide sequences with adapters.</p>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Updates self.orf_df with the final library DataFrame.
Updates self.file_name with the saved file path.</p>
<p>The CSV file contains columns:
- name: Variant identifier
- protein: Amino acid sequence
- orf: Nucleotide sequence (without adapters)
- oligo: Full oligonucleotide sequence (with adapters)</p>
<p>File naming convention:
{name}_orf_library_N={total_variants}_time={timestamp}.csv</p>
</div>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="DMS_designer.library_maker.main">
<span class="sig-prename descclassname"><span class="pre">DMS_designer.library_maker.</span></span><span class="sig-name descname"><span class="pre">main</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#main"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.library_maker.main" title="Permalink to this definition"></a></dt>
<dd><p>Main function for command-line interface.</p>
<p>This function provides the command-line interface for creating mutagenesis
libraries. It can be used to generate libraries from either nucleotide
or amino acid sequences.</p>
<div class="admonition-examples admonition">
<p class="admonition-title">Examples</p>
<p># Create library from nucleotide sequence
dms-maker -w ATGGCCGAA -s True -d equal -l even_single_and_double_nt_library</p>
<p># Create library from amino acid sequence
dms-maker -w MKEL -a -s True -d 100 -l even_single_and_double_aa_library</p>
<p># Create library with custom parameters
dms-maker -w ATGGCCGAA -s True -d max:10000 -n 5 -t 50 -l single_and_more_doubles_nt_library</p>
<p># Create library with focused mutations
dms-maker -w ATGGCCGAA -s True -d max:10000 -f 1 3 5 -c singles -p 3 -e 0.5 -l single_and_more_doubles_with_all_focused_mutations_nt_library</p>
</div>
</dd></dl>

</section>
</section>
<section id="module-DMS_designer">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-DMS_designer" title="Permalink to this heading"></a></h2>
<p>DMS Designer - Deep Mutational Scanning Library Design and Analysis Toolkit</p>
<p>A comprehensive toolkit for designing and analyzing deep mutational scanning (DMS) libraries.</p>
<p>This package provides tools for:
- Creating oligonucleotide libraries for protein mutagenesis studies
- Generating single and multiple amino acid mutants
- Creating synonymous nucleotide variants using codon usage tables
- Focused mutagenesis at specific positions
- Analyzing and visualizing library composition and results</p>
<p>Main Components:
- LibraryMaker: Main class for creating mutagenesis libraries
- LibraryAnalyzer: Tools for analyzing DMS library results
- Utility functions for codon/protein sequence conversion</p>
<dl>
<dt>Example Usage:</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">DMS_designer</span><span class="w"> </span><span class="kn">import</span> <span class="n">LibraryMaker</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lm</span> <span class="o">=</span> <span class="n">LibraryMaker</span><span class="p">(</span><span class="s2">&quot;ATGGCCGAA&quot;</span><span class="p">,</span> <span class="n">single_mutants</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">doubles_to_make</span><span class="o">=</span><span class="s1">&#39;equal&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">DMS_designer</span><span class="w"> </span><span class="kn">import</span> <span class="n">summarize_file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">summarize_file</span><span class="p">(</span><span class="s1">&#39;my_library.csv&#39;</span><span class="p">,</span> <span class="n">out_path</span><span class="o">=</span><span class="s1">&#39;./analysis/&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
<dl class="py class">
<dt class="sig sig-object py" id="DMS_designer.LibraryMaker">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">DMS_designer.</span></span><span class="sig-name descname"><span class="pre">LibraryMaker</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">wt_nt,</span> <span class="pre">single_mutants=True,</span> <span class="pre">doubles_to_make='equal',</span> <span class="pre">name='Mutant_lib',</span> <span class="pre">var_num=3,</span> <span class="pre">wt_num=100,</span> <span class="pre">max_tries=100,</span> <span class="pre">stop_num=None,</span> <span class="pre">focus_positions=[],</span> <span class="pre">focus_to_make=0,</span> <span class="pre">focus_mutants_per=3,</span> <span class="pre">focus_mutants_per_spread=0,</span> <span class="pre">focus_mutants_per_min=0,</span> <span class="pre">focus_mutants_per_max=0,</span> <span class="pre">upstream_dna='ATGGAAAGTGTCCCAGGAGACTACAGC',</span> <span class="pre">downstream_dna='CAGACAAGTAACTCAAGACTTAATGGC',</span> <span class="pre">out_path='./',</span> <span class="pre">seed=None,</span> <span class="pre">codon_df=</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">aa</span>&#160; <span class="pre">frac_of_aa</span>&#160; <span class="pre">ct_per_k</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">ct</span>&#160;&#160;&#160; <span class="pre">use</span> <span class="pre">codon</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">TAA</span>&#160;&#160;&#160; <span class="pre">*</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.61</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">2.0</span>&#160;&#160;&#160; <span class="pre">7356</span>&#160;&#160; <span class="pre">True</span> <span class="pre">TGA</span>&#160;&#160;&#160; <span class="pre">*</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.30</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">1.0</span>&#160;&#160;&#160; <span class="pre">3623</span>&#160;&#160; <span class="pre">True</span> <span class="pre">TAG</span>&#160;&#160;&#160; <span class="pre">*</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.09</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.3</span>&#160;&#160;&#160;&#160; <span class="pre">989</span>&#160; <span class="pre">False</span> <span class="pre">GCG</span>&#160;&#160;&#160; <span class="pre">A</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.33</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">30.1</span>&#160; <span class="pre">110308</span>&#160;&#160; <span class="pre">True</span> <span class="pre">GCC</span>&#160;&#160;&#160; <span class="pre">A</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.26</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">24.2</span>&#160;&#160; <span class="pre">88721</span>&#160;&#160; <span class="pre">True</span> <span class="pre">...</span>&#160;&#160; <span class="pre">..</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">...</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">...</span>&#160;&#160;&#160;&#160; <span class="pre">...</span>&#160;&#160;&#160; <span class="pre">...</span> <span class="pre">GTC</span>&#160;&#160;&#160; <span class="pre">V</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.20</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">14.3</span>&#160;&#160; <span class="pre">52439</span>&#160; <span class="pre">False</span> <span class="pre">GTA</span>&#160;&#160;&#160; <span class="pre">V</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.17</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">11.6</span>&#160;&#160; <span class="pre">42420</span>&#160; <span class="pre">False</span> <span class="pre">TGG</span>&#160;&#160;&#160; <span class="pre">W</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">1.00</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">13.9</span>&#160;&#160; <span class="pre">50991</span>&#160;&#160; <span class="pre">True</span> <span class="pre">TAT</span>&#160;&#160;&#160; <span class="pre">Y</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.59</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">17.5</span>&#160;&#160; <span class="pre">63937</span>&#160;&#160; <span class="pre">True</span> <span class="pre">TAC</span>&#160;&#160;&#160; <span class="pre">Y</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.41</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">12.2</span>&#160;&#160; <span class="pre">44631</span>&#160;&#160; <span class="pre">True</span>&#160; <span class="pre">[64</span> <span class="pre">rows</span> <span class="pre">x</span> <span class="pre">5</span> <span class="pre">columns],</span> <span class="pre">aa_to_codon_dict={'*':</span> <span class="pre">array(['TAA',</span> <span class="pre">'TGA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'A':</span> <span class="pre">array(['GCG',</span> <span class="pre">'GCC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'C':</span> <span class="pre">array(['TGC',</span> <span class="pre">'TGT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'D':</span> <span class="pre">array(['GAT',</span> <span class="pre">'GAC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'E':</span> <span class="pre">array(['GAA',</span> <span class="pre">'GAG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'F':</span> <span class="pre">array(['TTT',</span> <span class="pre">'TTC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'G':</span> <span class="pre">array(['GGC',</span> <span class="pre">'GGT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'H':</span> <span class="pre">array(['CAT',</span> <span class="pre">'CAC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'I':</span> <span class="pre">array(['ATT',</span> <span class="pre">'ATC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'K':</span> <span class="pre">array(['AAA',</span> <span class="pre">'AAG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'L':</span> <span class="pre">array(['CTG',</span> <span class="pre">'TTA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'M':</span> <span class="pre">array(['ATG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'N':</span> <span class="pre">array(['AAC',</span> <span class="pre">'AAT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'P':</span> <span class="pre">array(['CCG',</span> <span class="pre">'CCA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'Q':</span> <span class="pre">array(['CAG',</span> <span class="pre">'CAA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'R':</span> <span class="pre">array(['CGT',</span> <span class="pre">'CGC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'S':</span> <span class="pre">array(['AGC',</span> <span class="pre">'TCT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'T':</span> <span class="pre">array(['ACC',</span> <span class="pre">'ACG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'V':</span> <span class="pre">array(['GTG',</span> <span class="pre">'GTT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'W':</span> <span class="pre">array(['TGG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'Y':</span> <span class="pre">array(['TAT',</span> <span class="pre">'TAC'],</span> <span class="pre">dtype=object)},</span> <span class="pre">fig_width=8,</span> <span class="pre">context='talk',</span> <span class="pre">fig_path=None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#LibraryMaker"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.LibraryMaker" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Create oligonucleotide libraries for protein mutagenesis studies.</p>
<p>This class generates comprehensive mutagenesis libraries including single and
multiple amino acid mutants, synonymous nucleotide variants, and focused
mutagenesis at specific positions. The library is designed for deep mutational
scanning experiments.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wt_nt</strong> (<em>str</em>) – The nucleotide sequence of the wild-type variant. This sequence should
encode an amino acid sequence, start on the first nucleotide of the first
codon to be mutagenized, and end on the last nucleotide of the last
codon to be mutagenized.</p></li>
<li><p><strong>single_mutants</strong> (<em>bool</em><em>, </em><em>default True</em>) – If True, generate every possible single amino acid mutant of the wild-type
variant.</p></li>
<li><p><strong>doubles_to_make</strong> (<em>str</em><em>, </em><em>float</em><em>, or </em><em>int</em><em>, </em><em>default 'equal'</em>) – <p>Controls the number of double mutants to generate:</p>
<ul>
<li><p>’equal’: Generate random double mutants equal to the number of single mutants</p></li>
<li><p>’max:N’: Generate as many random double mutants as possible without exceeding
N total oligos for synthesis</p></li>
<li><p>float: Generate that fraction of the number of single mutants as random
double mutants</p></li>
<li><p>int: Generate exactly that number of random double mutants</p></li>
</ul>
</p></li>
<li><p><strong>name</strong> (<em>str</em><em>, </em><em>default 'Mutant_lib'</em>) – Name of the library to be created. Used as prefix for output files.</p></li>
<li><p><strong>var_num</strong> (<em>int</em><em>, </em><em>default 3</em>) – Number of different synonymous nucleotide variants to generate for each mutant.</p></li>
<li><p><strong>wt_num</strong> (<em>int</em><em>, </em><em>default 100</em>) – Number of different synonymous nucleotide variants to generate for the
wild-type sequence.</p></li>
<li><p><strong>max_tries</strong> (<em>int</em><em>, </em><em>default 100</em>) – Maximum number of attempts to generate a unique random mutant or variant
before giving up.</p></li>
<li><p><strong>focus_positions</strong> (<em>list of int</em><em>, </em><em>default</em><em> [</em><em>]</em>) – Amino acid positions (0-indexed) to be mutated in greater depth.</p></li>
<li><p><strong>focus_to_make</strong> (<em>int</em><em>, </em><em>float</em><em>, or </em><em>str</em><em>, </em><em>default 0</em>) – <p>Controls focused mutagenesis when focus_positions is specified:</p>
<ul>
<li><p>int/float: Generate that many random combinations of mutations at focus positions</p></li>
<li><p>’singles’: Generate all single mutants of the focus positions</p></li>
<li><p>’doubles’: Generate all double mutants of the focus positions</p></li>
<li><p>’singles_doubles’: Generate all single and double mutants of the focus positions</p></li>
<li><p>’all’: Generate every possible combination of mutations at the focus positions</p></li>
</ul>
</p></li>
<li><p><strong>focus_mutants_per</strong> (<em>int</em><em>, </em><em>default 3</em>) – Average number of mutations desired per focused mutant (used when focus_to_make
is numeric).</p></li>
<li><p><strong>focus_mutants_per_spread</strong> (<em>float</em><em>, </em><em>default 0</em>) – Probability spread for mutation numbers other than focus_mutants_per.
Probability decays arithmetically as you move away from the target.</p></li>
<li><p><strong>focus_mutants_per_min</strong> (<em>int</em><em>, </em><em>default 0</em>) – Minimum number of mutations per focused mutant.</p></li>
<li><p><strong>focus_mutants_per_max</strong> (<em>int</em><em>, </em><em>default 0</em>) – Maximum number of mutations per focused mutant.</p></li>
<li><p><strong>upstream_dna</strong> (<em>str</em><em>, </em><em>default 'ATGGAAAGTGTCCCAGGAGACTACAGC'</em>) – Upstream adapter sequence to append to the 5’ end of oligonucleotides.</p></li>
<li><p><strong>downstream_dna</strong> (<em>str</em><em>, </em><em>default 'CAGACAAGTAACTCAAGACTTAATGGC'</em>) – Downstream adapter sequence to append to the 3’ end of oligonucleotides.</p></li>
<li><p><strong>out_path</strong> (<em>str</em><em>, </em><em>default './'</em>) – Directory path for output files. Directory must already exist.</p></li>
<li><p><strong>seed</strong> (<em>int</em><em>, </em><em>optional</em>) – Random seed for reproducibility. If None, uses system default.</p></li>
<li><p><strong>fig_width</strong> (<em>int</em><em>, </em><em>default 8</em>) – Base width for generated analysis figures.</p></li>
<li><p><strong>context</strong> (<em>{'talk'</em><em>, </em><em>'paper'</em><em>, </em><em>'notebook'</em><em>, </em><em>'poster'}</em><em>, </em><em>default 'talk'</em>) – Seaborn plotting context for figure styling.</p></li>
</ul>
</dd>
<dt class="field-even">Variables<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>wt_nt</strong> (<em>str</em>) – Wild-type nucleotide sequence</p></li>
<li><p><strong>wt_protein</strong> (<em>str</em>) – Wild-type protein sequence</p></li>
<li><p><strong>L</strong> (<em>int</em>) – Length of the protein sequence</p></li>
<li><p><strong>variant_dict</strong> (<em>dict</em>) – Dictionary mapping variant names to protein sequences</p></li>
<li><p><strong>orf_dict</strong> (<em>dict</em>) – Dictionary mapping variant names to nucleotide sequences</p></li>
<li><p><strong>orf_df</strong> (<em>pandas.DataFrame</em>) – Final library data with columns: name, protein, orf, oligo</p></li>
<li><p><strong>file_name</strong> (<em>str</em>) – Name of the saved CSV file</p></li>
<li><p><strong>date_time</strong> (<em>str</em>) – Timestamp of library creation</p></li>
</ul>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>The library generation process:
1. Creates single mutants (if enabled)
2. Creates random double mutants (if enabled)
3. Creates focused mutants (if focus positions specified)
4. Generates synonymous nucleotide variants for each protein variant
5. Saves results to CSV file
6. Generates analysis plots and statistics</p>
</div>
<div class="admonition-examples admonition">
<p class="admonition-title">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># Create a simple library with single mutants only</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lm</span> <span class="o">=</span> <span class="n">LibraryMaker</span><span class="p">(</span><span class="s2">&quot;ATGGCCGAA&quot;</span><span class="p">,</span> <span class="n">single_mutants</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">doubles_to_make</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Create a comprehensive library with focused mutagenesis</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lm</span> <span class="o">=</span> <span class="n">LibraryMaker</span><span class="p">(</span><span class="s2">&quot;ATGGCCGAA&quot;</span><span class="p">,</span>
<span class="gp">... </span>                  <span class="n">single_mutants</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">... </span>                  <span class="n">doubles_to_make</span><span class="o">=</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span>
<span class="gp">... </span>                  <span class="n">focus_positions</span><span class="o">=</span><span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">15</span><span class="p">],</span>
<span class="gp">... </span>                  <span class="n">focus_to_make</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>This class creates oligo libraries for mutagenesis studies of protein function</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wt_nt</strong> (<em>string</em>) – The nucleotide sequence of the WT variant. This sequence should encode an AA sequence,
start on the first NT of the first codon to be mutagenized, and end on the last nt of the last codon to be mutagenized</p></li>
<li><p><strong>single_mutants</strong> (<em>bool Default is True</em>) – If true, make every possible single AA mutant of the WT variant</p></li>
<li><p><strong>doubles_to_make</strong> (<em>string</em><em>, </em><em>float</em><em>, or </em><em>int Default is 'equal</em>) – If this is the string ‘equal’ make random double mutants equal to the number of single mutants made
If this is a string starting with ‘max:’ follwed by an int, make as many random double mutants as possible
without allowing the number of oligos required to synthesize single and double mutants to exceed the number listed
If this is a float, Make that fraction of the number of single mutants as random duble mutants
If this is an int, Make that number of random double mutants</p></li>
<li><p><strong>name</strong> (<em>string Default is 'Mutant_lib'</em>) – name of the library to be created, will be appended as a prefix to the output files</p></li>
<li><p><strong>var_num</strong> (<em>int default 3</em>) – The number of different synonymous nt variants to be made for each mutant</p></li>
<li><p><strong>wt_num</strong> (<em>int default 100</em>) – The number of different synsonymous nt vatriants to be made for the WT sequence</p></li>
<li><p><strong>max_tries</strong> (<em>int default 100</em>) – The number of times to retry making a random mutant or variant if an already existing sequence was created</p></li>
<li><p><strong>stop_num</strong> (<em>int default None</em>) – The number of different synonymous nt variants to be made for each nonsense variant,
if None, default to var_num</p></li>
<li><p><strong>focus_positions</strong> (<em>list default</em><em> [</em><em>]</em>) – Amino acid positions of AAs to be mutated in greater depth</p></li>
<li><p><strong>focus_to_make</strong> (<em>int</em><em>, </em><em>float</em><em>, or </em><em>{'singles'</em><em>, </em><em>'doubles'</em><em>, </em><em>'singles_doubles'</em><em>, </em><em>'all'} default 0</em>) – This parameter is only used of focus_positions contains at least 1 element
If this is an int or a float, make that many random combinations of mutations at the focus positions
If this is the string ‘singles’, make all single mutants of the focus positions
If this is the string ‘doubles’, make all double mutants of the focus positions
If this is the string ‘singles_doubles’, make all single and double muatants of the focus positions
If this is the string ‘all’, make every combination of mutations at the focus positions</p></li>
<li><p><strong>focus_mutants_per</strong> (<em>int default 0</em>) – This parameter is only used of focus_positions contains at least 1 element and focus_to_make is an int or float &gt;0
this indicates the average number of mutations desired per mutant</p></li>
<li><p><strong>focus_mutants_per_spread</strong> (<em>int default 0</em>) – This parameter is only used of focus_positions contains at least 1 element and focus_to_make is an int or float &gt;0
This is the probability of producing a mutation number other than focus_mutants_per, mutation probability will be evenly distributed
between higher and lower mutation numbers with arithmatic decay as you move away from focus_mutants_per</p></li>
<li><p><strong>focus_mutants_per_min</strong> – This parameter is only used of focus_positions contains at least 1 element and focus_to_make is an int or float &gt;0
This is the minimum mutation number you want to see per mutant</p></li>
<li><p><strong>focus_mutants_per_max</strong> – This parameter is only used of focus_positions contains at least 1 element and focus_to_make is an int or float &gt;0
This is the Maximum mutation number you want to see per mutant</p></li>
<li><p><strong>upstream_dna</strong> (<em>string default 'ATGGAAAGTGTCCCAGGAGACTACAGC'</em>) – This is the upstream adaptor sequence to append to the 5’ end of oligos</p></li>
<li><p><strong>downstream_dna</strong> (<em>string default 'CAGACAAGTAACTCAAGACTTAATGGC'</em>) – This is the downstream adaptor sequence to append to the 3’ end of oligos</p></li>
<li><p><strong>out_path</strong> (<em>string default './'</em>) – path to place output files. This directory must already exist</p></li>
<li><p><strong>codon_df</strong> (<em>pd.DataFrame</em><em>, </em><em>optional</em>) – DataFrame containing codon usage information. If None, uses the default codon usage table.</p></li>
<li><p><strong>aa_to_codon_dict</strong> (<em>dict</em><em>, </em><em>optional</em>) – Dictionary mapping amino acids to their codons. If None, uses the default codon usage table.</p></li>
<li><p><strong>seed</strong> (<em>int Default None</em>) – Random seed. Set to ensure reproducibility.</p></li>
<li><p><strong>fig_width</strong> (<em>int default 8</em>) – Size parameter for output figures</p></li>
<li><p><strong>context</strong> (<em>{'talk'</em><em>, </em><em>'paper'</em><em>, </em><em>'notebook'</em><em>, </em><em>'poster'} default 'talk'</em>) – Used to set plotting configuration. passed to seaborn.set_context</p></li>
<li><p><strong>fig_path</strong> (<em>str</em><em>, </em><em>default None</em>) – Path to save the output figures. If default to out_path.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.LibraryMaker.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">wt_nt,</span> <span class="pre">single_mutants=True,</span> <span class="pre">doubles_to_make='equal',</span> <span class="pre">name='Mutant_lib',</span> <span class="pre">var_num=3,</span> <span class="pre">wt_num=100,</span> <span class="pre">max_tries=100,</span> <span class="pre">stop_num=None,</span> <span class="pre">focus_positions=[],</span> <span class="pre">focus_to_make=0,</span> <span class="pre">focus_mutants_per=3,</span> <span class="pre">focus_mutants_per_spread=0,</span> <span class="pre">focus_mutants_per_min=0,</span> <span class="pre">focus_mutants_per_max=0,</span> <span class="pre">upstream_dna='ATGGAAAGTGTCCCAGGAGACTACAGC',</span> <span class="pre">downstream_dna='CAGACAAGTAACTCAAGACTTAATGGC',</span> <span class="pre">out_path='./',</span> <span class="pre">seed=None,</span> <span class="pre">codon_df=</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">aa</span>&#160; <span class="pre">frac_of_aa</span>&#160; <span class="pre">ct_per_k</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">ct</span>&#160;&#160;&#160; <span class="pre">use</span> <span class="pre">codon</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">TAA</span>&#160;&#160;&#160; <span class="pre">*</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.61</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">2.0</span>&#160;&#160;&#160; <span class="pre">7356</span>&#160;&#160; <span class="pre">True</span> <span class="pre">TGA</span>&#160;&#160;&#160; <span class="pre">*</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.30</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">1.0</span>&#160;&#160;&#160; <span class="pre">3623</span>&#160;&#160; <span class="pre">True</span> <span class="pre">TAG</span>&#160;&#160;&#160; <span class="pre">*</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.09</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.3</span>&#160;&#160;&#160;&#160; <span class="pre">989</span>&#160; <span class="pre">False</span> <span class="pre">GCG</span>&#160;&#160;&#160; <span class="pre">A</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.33</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">30.1</span>&#160; <span class="pre">110308</span>&#160;&#160; <span class="pre">True</span> <span class="pre">GCC</span>&#160;&#160;&#160; <span class="pre">A</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.26</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">24.2</span>&#160;&#160; <span class="pre">88721</span>&#160;&#160; <span class="pre">True</span> <span class="pre">...</span>&#160;&#160; <span class="pre">..</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">...</span>&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">...</span>&#160;&#160;&#160;&#160; <span class="pre">...</span>&#160;&#160;&#160; <span class="pre">...</span> <span class="pre">GTC</span>&#160;&#160;&#160; <span class="pre">V</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.20</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">14.3</span>&#160;&#160; <span class="pre">52439</span>&#160; <span class="pre">False</span> <span class="pre">GTA</span>&#160;&#160;&#160; <span class="pre">V</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.17</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">11.6</span>&#160;&#160; <span class="pre">42420</span>&#160; <span class="pre">False</span> <span class="pre">TGG</span>&#160;&#160;&#160; <span class="pre">W</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">1.00</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">13.9</span>&#160;&#160; <span class="pre">50991</span>&#160;&#160; <span class="pre">True</span> <span class="pre">TAT</span>&#160;&#160;&#160; <span class="pre">Y</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.59</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">17.5</span>&#160;&#160; <span class="pre">63937</span>&#160;&#160; <span class="pre">True</span> <span class="pre">TAC</span>&#160;&#160;&#160; <span class="pre">Y</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">0.41</span>&#160;&#160;&#160;&#160;&#160; <span class="pre">12.2</span>&#160;&#160; <span class="pre">44631</span>&#160;&#160; <span class="pre">True</span>&#160; <span class="pre">[64</span> <span class="pre">rows</span> <span class="pre">x</span> <span class="pre">5</span> <span class="pre">columns],</span> <span class="pre">aa_to_codon_dict={'*':</span> <span class="pre">array(['TAA',</span> <span class="pre">'TGA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'A':</span> <span class="pre">array(['GCG',</span> <span class="pre">'GCC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'C':</span> <span class="pre">array(['TGC',</span> <span class="pre">'TGT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'D':</span> <span class="pre">array(['GAT',</span> <span class="pre">'GAC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'E':</span> <span class="pre">array(['GAA',</span> <span class="pre">'GAG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'F':</span> <span class="pre">array(['TTT',</span> <span class="pre">'TTC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'G':</span> <span class="pre">array(['GGC',</span> <span class="pre">'GGT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'H':</span> <span class="pre">array(['CAT',</span> <span class="pre">'CAC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'I':</span> <span class="pre">array(['ATT',</span> <span class="pre">'ATC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'K':</span> <span class="pre">array(['AAA',</span> <span class="pre">'AAG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'L':</span> <span class="pre">array(['CTG',</span> <span class="pre">'TTA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'M':</span> <span class="pre">array(['ATG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'N':</span> <span class="pre">array(['AAC',</span> <span class="pre">'AAT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'P':</span> <span class="pre">array(['CCG',</span> <span class="pre">'CCA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'Q':</span> <span class="pre">array(['CAG',</span> <span class="pre">'CAA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'R':</span> <span class="pre">array(['CGT',</span> <span class="pre">'CGC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'S':</span> <span class="pre">array(['AGC',</span> <span class="pre">'TCT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'T':</span> <span class="pre">array(['ACC',</span> <span class="pre">'ACG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'V':</span> <span class="pre">array(['GTG',</span> <span class="pre">'GTT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'W':</span> <span class="pre">array(['TGG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'Y':</span> <span class="pre">array(['TAT',</span> <span class="pre">'TAC'],</span> <span class="pre">dtype=object)},</span> <span class="pre">fig_width=8,</span> <span class="pre">context='talk',</span> <span class="pre">fig_path=None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#LibraryMaker.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.LibraryMaker.__init__" title="Permalink to this definition"></a></dt>
<dd><p>This class creates oligo libraries for mutagenesis studies of protein function</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>wt_nt</strong> (<em>string</em>) – The nucleotide sequence of the WT variant. This sequence should encode an AA sequence,
start on the first NT of the first codon to be mutagenized, and end on the last nt of the last codon to be mutagenized</p></li>
<li><p><strong>single_mutants</strong> (<em>bool Default is True</em>) – If true, make every possible single AA mutant of the WT variant</p></li>
<li><p><strong>doubles_to_make</strong> (<em>string</em><em>, </em><em>float</em><em>, or </em><em>int Default is 'equal</em>) – If this is the string ‘equal’ make random double mutants equal to the number of single mutants made
If this is a string starting with ‘max:’ follwed by an int, make as many random double mutants as possible
without allowing the number of oligos required to synthesize single and double mutants to exceed the number listed
If this is a float, Make that fraction of the number of single mutants as random duble mutants
If this is an int, Make that number of random double mutants</p></li>
<li><p><strong>name</strong> (<em>string Default is 'Mutant_lib'</em>) – name of the library to be created, will be appended as a prefix to the output files</p></li>
<li><p><strong>var_num</strong> (<em>int default 3</em>) – The number of different synonymous nt variants to be made for each mutant</p></li>
<li><p><strong>wt_num</strong> (<em>int default 100</em>) – The number of different synsonymous nt vatriants to be made for the WT sequence</p></li>
<li><p><strong>max_tries</strong> (<em>int default 100</em>) – The number of times to retry making a random mutant or variant if an already existing sequence was created</p></li>
<li><p><strong>stop_num</strong> (<em>int default None</em>) – The number of different synonymous nt variants to be made for each nonsense variant,
if None, default to var_num</p></li>
<li><p><strong>focus_positions</strong> (<em>list default</em><em> [</em><em>]</em>) – Amino acid positions of AAs to be mutated in greater depth</p></li>
<li><p><strong>focus_to_make</strong> (<em>int</em><em>, </em><em>float</em><em>, or </em><em>{'singles'</em><em>, </em><em>'doubles'</em><em>, </em><em>'singles_doubles'</em><em>, </em><em>'all'} default 0</em>) – This parameter is only used of focus_positions contains at least 1 element
If this is an int or a float, make that many random combinations of mutations at the focus positions
If this is the string ‘singles’, make all single mutants of the focus positions
If this is the string ‘doubles’, make all double mutants of the focus positions
If this is the string ‘singles_doubles’, make all single and double muatants of the focus positions
If this is the string ‘all’, make every combination of mutations at the focus positions</p></li>
<li><p><strong>focus_mutants_per</strong> (<em>int default 0</em>) – This parameter is only used of focus_positions contains at least 1 element and focus_to_make is an int or float &gt;0
this indicates the average number of mutations desired per mutant</p></li>
<li><p><strong>focus_mutants_per_spread</strong> (<em>int default 0</em>) – This parameter is only used of focus_positions contains at least 1 element and focus_to_make is an int or float &gt;0
This is the probability of producing a mutation number other than focus_mutants_per, mutation probability will be evenly distributed
between higher and lower mutation numbers with arithmatic decay as you move away from focus_mutants_per</p></li>
<li><p><strong>focus_mutants_per_min</strong> – This parameter is only used of focus_positions contains at least 1 element and focus_to_make is an int or float &gt;0
This is the minimum mutation number you want to see per mutant</p></li>
<li><p><strong>focus_mutants_per_max</strong> – This parameter is only used of focus_positions contains at least 1 element and focus_to_make is an int or float &gt;0
This is the Maximum mutation number you want to see per mutant</p></li>
<li><p><strong>upstream_dna</strong> (<em>string default 'ATGGAAAGTGTCCCAGGAGACTACAGC'</em>) – This is the upstream adaptor sequence to append to the 5’ end of oligos</p></li>
<li><p><strong>downstream_dna</strong> (<em>string default 'CAGACAAGTAACTCAAGACTTAATGGC'</em>) – This is the downstream adaptor sequence to append to the 3’ end of oligos</p></li>
<li><p><strong>out_path</strong> (<em>string default './'</em>) – path to place output files. This directory must already exist</p></li>
<li><p><strong>codon_df</strong> (<em>pd.DataFrame</em><em>, </em><em>optional</em>) – DataFrame containing codon usage information. If None, uses the default codon usage table.</p></li>
<li><p><strong>aa_to_codon_dict</strong> (<em>dict</em><em>, </em><em>optional</em>) – Dictionary mapping amino acids to their codons. If None, uses the default codon usage table.</p></li>
<li><p><strong>seed</strong> (<em>int Default None</em>) – Random seed. Set to ensure reproducibility.</p></li>
<li><p><strong>fig_width</strong> (<em>int default 8</em>) – Size parameter for output figures</p></li>
<li><p><strong>context</strong> (<em>{'talk'</em><em>, </em><em>'paper'</em><em>, </em><em>'notebook'</em><em>, </em><em>'poster'} default 'talk'</em>) – Used to set plotting configuration. passed to seaborn.set_context</p></li>
<li><p><strong>fig_path</strong> (<em>str</em><em>, </em><em>default None</em>) – Path to save the output figures. If default to out_path.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.LibraryMaker.create_focused_singles">
<span class="sig-name descname"><span class="pre">create_focused_singles</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#LibraryMaker.create_focused_singles"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.LibraryMaker.create_focused_singles" title="Permalink to this definition"></a></dt>
<dd><p>Create all single amino acid mutants at focused positions.</p>
<p>Generates every possible single amino acid substitution at each position
specified in focus_positions, excluding the wild-type amino acid.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary mapping variant names to protein sequences for focused
single mutants. Names follow pattern 1pt_PPAA where PP is the
position (01-indexed) and AA is the amino acid.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Updates self.variant_dict with the new variants.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.LibraryMaker.create_focused_doubles">
<span class="sig-name descname"><span class="pre">create_focused_doubles</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#LibraryMaker.create_focused_doubles"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.LibraryMaker.create_focused_doubles" title="Permalink to this definition"></a></dt>
<dd><p>Create all double amino acid mutants at focused positions.</p>
<p>Generates every possible combination of two amino acid substitutions
at pairs of positions specified in focus_positions, excluding wild-type
amino acids and stop codons.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary mapping variant names to protein sequences for focused
double mutants. Names follow pattern 2pt_PP1AA1_PP2AA2 where PP
is the position (01-indexed) and AA is the amino acid.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Updates self.variant_dict with the new variants.
Excludes stop codons (*) from possible substitutions.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.LibraryMaker.create_focused_all_combinations">
<span class="sig-name descname"><span class="pre">create_focused_all_combinations</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#LibraryMaker.create_focused_all_combinations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.LibraryMaker.create_focused_all_combinations" title="Permalink to this definition"></a></dt>
<dd><p>Create all possible combinations of mutations at focused positions.</p>
<p>Generates every possible combination of amino acid substitutions at the
focused positions, from single mutations up to mutations at all focused
positions simultaneously. Excludes wild-type amino acids and stop codons.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary mapping variant names to protein sequences for all focused
mutant combinations. Names follow pattern <a href="#id5"><span class="problematic" id="id6">Npt_PP1AA1_PP2AA2_</span></a>…
where N is the number of mutations, PP is position (01-indexed),
and AA is the amino acid.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Updates self.variant_dict with the new variants.
Excludes stop codons (*) from possible substitutions.
Generates combinations of all sizes from 1 to len(focus_positions).</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.LibraryMaker.create_focused_random_combinations">
<span class="sig-name descname"><span class="pre">create_focused_random_combinations</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#LibraryMaker.create_focused_random_combinations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.LibraryMaker.create_focused_random_combinations" title="Permalink to this definition"></a></dt>
<dd><p>Create random combinations of mutations at focused positions.</p>
<p>Generates random combinations of amino acid substitutions at focused
positions according to specified probability distributions and constraints.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary mapping variant names to protein sequences for random
focused mutant combinations. Names follow pattern <a href="#id7"><span class="problematic" id="id8">Npt_PP1AA1_PP2AA2_</span></a>…
where N is the number of mutations, PP is position (01-indexed),
and AA is the amino acid.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Updates self.variant_dict with the new variants.</p>
<p>The number of mutations per variant is controlled by:
- focus_mutants_per: Target average number of mutations
- focus_mutants_per_spread: Probability spread around the target
- focus_mutants_per_min: Minimum mutations per variant
- focus_mutants_per_max: Maximum mutations per variant</p>
<p>If spread &gt; 0, probability decays arithmetically as you move away
from the target number of mutations.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.LibraryMaker.create_singles">
<span class="sig-name descname"><span class="pre">create_singles</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#LibraryMaker.create_singles"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.LibraryMaker.create_singles" title="Permalink to this definition"></a></dt>
<dd><p>Create all possible single amino acid mutants.</p>
<p>Generates every possible single amino acid substitution at each position
in the protein sequence, excluding the wild-type amino acid.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary mapping variant names to protein sequences for single
mutants. Names follow pattern 1pt_PPAA where PP is the position
(01-indexed) and AA is the amino acid.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Updates self.variant_dict with the new variants.
Creates (L * 20) single mutants where L is the protein length.
The 20 includes the stop codons.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.LibraryMaker.create_doubles">
<span class="sig-name descname"><span class="pre">create_doubles</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#LibraryMaker.create_doubles"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.LibraryMaker.create_doubles" title="Permalink to this definition"></a></dt>
<dd><p>Create random double amino acid mutants.</p>
<p>Generates random combinations of two amino acid substitutions at different
positions in the protein sequence. Uses a tracking matrix to avoid
generating duplicate combinations.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary mapping variant names to protein sequences for random
double mutants. Names follow pattern 2pt_PP1AA1_PP2AA2 where PP
is the position (01-indexed) and AA is the amino acid.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Updates self.variant_dict with the new variants.
Uses self.already_seen matrix to track which combinations have been
generated to avoid duplicates.
Excludes stop codons (*) from possible substitutions.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.LibraryMaker.generate_nt_seqs">
<span class="sig-name descname"><span class="pre">generate_nt_seqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">aa_to_codon_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'*':</span> <span class="pre">array(['TAA',</span> <span class="pre">'TGA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'A':</span> <span class="pre">array(['GCG',</span> <span class="pre">'GCC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'C':</span> <span class="pre">array(['TGC',</span> <span class="pre">'TGT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'D':</span> <span class="pre">array(['GAT',</span> <span class="pre">'GAC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'E':</span> <span class="pre">array(['GAA',</span> <span class="pre">'GAG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'F':</span> <span class="pre">array(['TTT',</span> <span class="pre">'TTC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'G':</span> <span class="pre">array(['GGC',</span> <span class="pre">'GGT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'H':</span> <span class="pre">array(['CAT',</span> <span class="pre">'CAC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'I':</span> <span class="pre">array(['ATT',</span> <span class="pre">'ATC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'K':</span> <span class="pre">array(['AAA',</span> <span class="pre">'AAG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'L':</span> <span class="pre">array(['CTG',</span> <span class="pre">'TTA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'M':</span> <span class="pre">array(['ATG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'N':</span> <span class="pre">array(['AAC',</span> <span class="pre">'AAT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'P':</span> <span class="pre">array(['CCG',</span> <span class="pre">'CCA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'Q':</span> <span class="pre">array(['CAG',</span> <span class="pre">'CAA'],</span> <span class="pre">dtype=object),</span> <span class="pre">'R':</span> <span class="pre">array(['CGT',</span> <span class="pre">'CGC'],</span> <span class="pre">dtype=object),</span> <span class="pre">'S':</span> <span class="pre">array(['AGC',</span> <span class="pre">'TCT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'T':</span> <span class="pre">array(['ACC',</span> <span class="pre">'ACG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'V':</span> <span class="pre">array(['GTG',</span> <span class="pre">'GTT'],</span> <span class="pre">dtype=object),</span> <span class="pre">'W':</span> <span class="pre">array(['TGG'],</span> <span class="pre">dtype=object),</span> <span class="pre">'Y':</span> <span class="pre">array(['TAT',</span> <span class="pre">'TAC'],</span> <span class="pre">dtype=object)}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#LibraryMaker.generate_nt_seqs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.LibraryMaker.generate_nt_seqs" title="Permalink to this definition"></a></dt>
<dd><p>Generate synonymous nucleotide sequences for all protein variants.</p>
<p>Creates multiple synonymous nucleotide sequences for each protein variant
using the codon usage table. The number of variants depends on whether
it’s the wild-type (wt_num variants) or a mutant (var_num variants).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>aa_to_codon_dict</strong> (<em>dict</em><em>, </em><em>optional</em>) – Dictionary mapping amino acids to their codons. If None, uses the default codon usage table.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary mapping variant names to nucleotide sequences.
Names follow pattern PROTEIN_NAME_vN where N is the variant number.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Updates self.orf_dict with the nucleotide sequences.</p>
<p>For wild-type variants, generates wt_num synonymous sequences.
For mutant variants, generates var_num synonymous sequences.</p>
<p>The first wild-type variant uses the exact input nucleotide sequence.
Subsequent variants are generated by randomly selecting one of the two most common codons for each aa.</p>
<p>Attempts up to max_tries times to generate unique sequences before
giving up on a particular variant.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.LibraryMaker.save_file">
<span class="sig-name descname"><span class="pre">save_file</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#LibraryMaker.save_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.LibraryMaker.save_file" title="Permalink to this definition"></a></dt>
<dd><p>Save the library to a CSV file and generate oligonucleotide sequences.</p>
<p>Creates a pandas DataFrame with all library variants and saves it to
a CSV file. The DataFrame includes protein sequences, nucleotide
sequences, and full oligonucleotide sequences with adapters.</p>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>Updates self.orf_df with the final library DataFrame.
Updates self.file_name with the saved file path.</p>
<p>The CSV file contains columns:
- name: Variant identifier
- protein: Amino acid sequence
- orf: Nucleotide sequence (without adapters)
- oligo: Full oligonucleotide sequence (with adapters)</p>
<p>File naming convention:
{name}_orf_library_N={total_variants}_time={timestamp}.csv</p>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="DMS_designer.Library">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">DMS_designer.</span></span><span class="sig-name descname"><span class="pre">Library</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_cols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_cols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rate_method</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alphabet</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">WT_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fitness_measure</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_cols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_cols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rate_method</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alphabet</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">WT_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fitness_measure</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.__init__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.save"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.save" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.load">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.load"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.load" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.build_from_reads">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">build_from_reads</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_cols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_cols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rate_method</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alphabet</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">WT_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">process_call</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name_regex</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sheet_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.build_from_reads"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.build_from_reads" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.merge_by_overlap">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">merge_by_overlap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">library_dest</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">library_source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dest_position_shift</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_position_shift</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_WT</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Position'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">automatically_rescale_linearly</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">automatically_rescale_by_controls</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'AA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_val</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['WT',</span> <span class="pre">'*']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.merge_by_overlap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.merge_by_overlap" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="DMS_designer.Library.fitness_frame">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">fitness_frame</span></span><a class="headerlink" href="#DMS_designer.Library.fitness_frame" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="DMS_designer.Library.rep_average_frame">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">rep_average_frame</span></span><a class="headerlink" href="#DMS_designer.Library.rep_average_frame" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="DMS_designer.Library.group_average_frame">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">group_average_frame</span></span><a class="headerlink" href="#DMS_designer.Library.group_average_frame" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="DMS_designer.Library.total_average_frame">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">total_average_frame</span></span><a class="headerlink" href="#DMS_designer.Library.total_average_frame" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.check_skew">
<span class="sig-name descname"><span class="pre">check_skew</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sns_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'notebook'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rc_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Library</span> <span class="pre">skew'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.check_skew"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.check_skew" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.check_depth">
<span class="sig-name descname"><span class="pre">check_depth</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sns_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'notebook'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rc_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Read</span> <span class="pre">depth'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.check_depth"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.check_depth" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.check_barcode_distribution_skew">
<span class="sig-name descname"><span class="pre">check_barcode_distribution_skew</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sns_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'notebook'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rc_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Read</span> <span class="pre">depth'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.check_barcode_distribution_skew"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.check_barcode_distribution_skew" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.check_barcode_distribution_depth">
<span class="sig-name descname"><span class="pre">check_barcode_distribution_depth</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sns_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'notebook'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rc_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Read</span> <span class="pre">depth'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.check_barcode_distribution_depth"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.check_barcode_distribution_depth" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.check_controls">
<span class="sig-name descname"><span class="pre">check_controls</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">control_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">positive_vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">negative_vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sns_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'notebook'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rc_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plt_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'strip'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_hist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.check_controls"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.check_controls" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.make_replicate_checks">
<span class="sig-name descname"><span class="pre">make_replicate_checks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sns_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'notebook'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rc_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">grid_kwargs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.make_replicate_checks"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.make_replicate_checks" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.get_rescaler_by_replicate">
<span class="sig-name descname"><span class="pre">get_rescaler_by_replicate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dest_replicate</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_replicate</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.get_rescaler_by_replicate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.get_rescaler_by_replicate" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.rescale_by_replicate">
<span class="sig-name descname"><span class="pre">rescale_by_replicate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dest_replicate</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_replicate</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.rescale_by_replicate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.rescale_by_replicate" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.get_rescaler_by_replicate_groups">
<span class="sig-name descname"><span class="pre">get_rescaler_by_replicate_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dest_replicate</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_replicate</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.get_rescaler_by_replicate_groups"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.get_rescaler_by_replicate_groups" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.rescale_by_replicate_group">
<span class="sig-name descname"><span class="pre">rescale_by_replicate_group</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dest_replicate</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_replicate</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.rescale_by_replicate_group"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.rescale_by_replicate_group" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.get_multi_replicate_rescaler_by_groups">
<span class="sig-name descname"><span class="pre">get_multi_replicate_rescaler_by_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dest_replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_replicates</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.get_multi_replicate_rescaler_by_groups"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.get_multi_replicate_rescaler_by_groups" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.rescale_multiple_replicates_by_group">
<span class="sig-name descname"><span class="pre">rescale_multiple_replicates_by_group</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dest_replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_replicates</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.rescale_multiple_replicates_by_group"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.rescale_multiple_replicates_by_group" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.get_multi_replicate_rescaler_by_controls">
<span class="sig-name descname"><span class="pre">get_multi_replicate_rescaler_by_controls</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dest_replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.get_multi_replicate_rescaler_by_controls"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.get_multi_replicate_rescaler_by_controls" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.rescale_multiple_replicates_by_controls">
<span class="sig-name descname"><span class="pre">rescale_multiple_replicates_by_controls</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dest_replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.rescale_multiple_replicates_by_controls"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.rescale_multiple_replicates_by_controls" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.make_initial_skew_checks">
<span class="sig-name descname"><span class="pre">make_initial_skew_checks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sns_context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'notebook'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rc_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.make_initial_skew_checks"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.make_initial_skew_checks" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.get_single_mutant_mutability_score">
<span class="sig-name descname"><span class="pre">get_single_mutant_mutability_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">control_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">WT_vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">thresh</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">invert</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.get_single_mutant_mutability_score"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.get_single_mutant_mutability_score" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.get_single_mutant_property_enrichments">
<span class="sig-name descname"><span class="pre">get_single_mutant_property_enrichments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">property_sets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">thresh</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Position'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">char_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'AA'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.get_single_mutant_property_enrichments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.get_single_mutant_property_enrichments" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.draw_singles_heatmap">
<span class="sig-name descname"><span class="pre">draw_singles_heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mut_num_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">WT_vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">grid</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tick_interval</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_pos</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.draw_singles_heatmap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.draw_singles_heatmap" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.draw_mutability_score">
<span class="sig-name descname"><span class="pre">draw_mutability_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">WT_vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_vals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">thresh</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">invert_mutability</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">grid</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_pos</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.draw_mutability_score"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.draw_mutability_score" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.draw_property_scores">
<span class="sig-name descname"><span class="pre">draw_property_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">property_sets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">thresh</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">grid</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_pos</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.draw_property_scores"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.draw_property_scores" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.draw_feature_sets">
<span class="sig-name descname"><span class="pre">draw_feature_sets</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">feature_sets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">grid</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_pos</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_multi_rows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.draw_feature_sets"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.draw_feature_sets" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="DMS_designer.Library.make_singles_heatmap">
<span class="sig-name descname"><span class="pre">make_singles_heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">mut_num_col='Mut_num',</span> <span class="pre">position_col='Position',</span> <span class="pre">main_cmap='RdBu',</span> <span class="pre">mutability_cmap=&lt;matplotlib.colors.LinearSegmentedColormap</span> <span class="pre">object&gt;,</span> <span class="pre">bad_color='xkcd:grey',</span> <span class="pre">sns_context='notebook',</span> <span class="pre">rc_params={},</span> <span class="pre">grid=False,</span> <span class="pre">invert_mutability=False,</span> <span class="pre">norm=None,</span> <span class="pre">seq_start=0,</span> <span class="pre">tick_interval=5,</span> <span class="pre">plot_pos=None,</span> <span class="pre">plot_mutability=False,</span> <span class="pre">plot_properties=False,</span> <span class="pre">plot_features=False,</span> <span class="pre">mutability_norm=None,</span> <span class="pre">property_norm=None,</span> <span class="pre">property_sets={'Charged':</span> <span class="pre">['R',</span> <span class="pre">'H',</span> <span class="pre">'K',</span> <span class="pre">'D',</span> <span class="pre">'E'],</span> <span class="pre">'Hydrophobic':</span> <span class="pre">['A',</span> <span class="pre">'V',</span> <span class="pre">'I',</span> <span class="pre">'L',</span> <span class="pre">'M',</span> <span class="pre">'F',</span> <span class="pre">'Y',</span> <span class="pre">'W'],</span> <span class="pre">'Polar':</span> <span class="pre">['S',</span> <span class="pre">'T',</span> <span class="pre">'N',</span> <span class="pre">'Q']},</span> <span class="pre">property_cmap='RdBu',</span> <span class="pre">feature_sets=[],</span> <span class="pre">control_col='AA',</span> <span class="pre">WT_vals=['WT'],</span> <span class="pre">stop_vals=['*'],</span> <span class="pre">thresh=0,</span> <span class="pre">allow_multi_rows=False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_analyzer.html#Library.make_singles_heatmap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.Library.make_singles_heatmap" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="DMS_designer.summarize_file">
<span class="sig-prename descclassname"><span class="pre">DMS_designer.</span></span><span class="sig-name descname"><span class="pre">summarize_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'./'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fig_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">12</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">context</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'talk'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/DMS_designer/library_maker.html#summarize_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#DMS_designer.summarize_file" title="Permalink to this definition"></a></dt>
<dd><p>Generate comprehensive analysis and visualization of a mutagenesis library file.</p>
<p>This function reads a CSV file containing library variants and generates multiple
plots and statistics including variant counts, mutant distributions, heatmaps,
and position-by-position analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> (<em>str</em>) – Path to the CSV file containing the library data. Expected columns:
‘name’, ‘protein’, ‘orf’, ‘oligo’</p></li>
<li><p><strong>out_path</strong> (<em>str</em><em>, </em><em>default './'</em>) – Directory path where output figures will be saved</p></li>
<li><p><strong>fig_width</strong> (<em>int</em><em>, </em><em>default 12</em>) – Base width for generated figures. Height is typically equal to width</p></li>
<li><p><strong>context</strong> (<em>{'talk'</em><em>, </em><em>'paper'</em><em>, </em><em>'notebook'</em><em>, </em><em>'poster'}</em><em>, </em><em>default 'talk'</em>) – Seaborn plotting context for figure styling</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
<div class="admonition-notes admonition">
<p class="admonition-title">Notes</p>
<p>The function generates the following output files:
- variant_count_[filename].pdf: Count of variants by mutation number
- mutant_count_[filename].pdf: Count of unique mutants by mutation number
- mutant_fraction_[filename].pdf: Fraction of possible mutants generated
- mutant_observations_[filename].pdf: Distribution of observations per mutant
- [N]pt_heatmap_[filename].pdf: Heatmaps showing amino acid frequencies at each position
- [N]pt_mutant_heatmap_[filename].pdf: Heatmaps for unique mutants only
- 2pt_2d_heatmap_[filename].pdf: Position-by-position matrix for double mutants</p>
<p>The function also prints summary statistics to the console including:
- Wild-type sequence information
- Counts of variants by mutation number
- Total library size
- Fraction of possible mutants generated</p>
</div>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="modules.html" class="btn btn-neutral float-left" title="DMS_designer" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, John J Desmarais.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>